{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Survival analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we examine whether CXCR6 expression predicts survival in human cancer patients using bulk RNA-Seq data sets publically available from the TCGA pan-cancer dataset: https://gdc.cancer.gov/about-data/publications/pancanatlas. Fig. 7 & S7 were generated using the code in this notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from scipy.stats import spearmanr\n",
    "from scipy.stats import f_oneway\n",
    "from scipy import stats\n",
    "from scipy.stats import zscore\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from collections import Counter\n",
    "\n",
    "from lifelines import KaplanMeierFitter\n",
    "from lifelines.statistics import logrank_test, multivariate_logrank_test\n",
    "from lifelines.plotting import add_at_risk_counts\n",
    "from lifelines import CoxPHFitter\n",
    "from lifelines import CoxTimeVaryingFitter\n",
    "\n",
    "import mm_functions_14_10_2020 as mm\n",
    "\n",
    "import itertools\n",
    "import functools\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set figure parameters \n",
    "mylines = 0.15*2.82 # the number 2.82 is the difference\n",
    "\t\t\t\t\t# between Illustrator 1 pt and python 1 pt.\n",
    "mpl.rcParams['axes.linewidth'] = mylines # default 1\n",
    "mpl.rcParams['ytick.direction'] = 'out' # default 'in'\n",
    "mpl.rcParams['xtick.direction'] = 'out' # default 'in'\n",
    "mpl.rcParams['xtick.major.size'] = 2 # default 4\n",
    "mpl.rcParams['ytick.major.size'] = 2 # default 4\n",
    "mpl.rcParams['xtick.major.width'] = mylines # default 0.5\n",
    "mpl.rcParams['ytick.major.width'] = mylines # default 0.5\n",
    "mpl.rcParams['grid.linewidth'] = mylines/1.5 # default 0.5\n",
    "mpl.rcParams['grid.color'] = '0.8' # default 'k'\n",
    "mpl.rcParams['grid.linestyle'] = 'solid'# default ':'\n",
    "mpl.rcParams['legend.frameon'] = False # default True\n",
    "mpl.rcParams['figure.dpi']= 300\n",
    "mpl.rc(\"savefig\", dpi=300)\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42\n",
    "plt.rc('font', family = 'Arial',size = 12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Survival functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fit_cph(end1_name = '', end1_time = None, end1_event = None, \n",
    "            end2_name = '', end2_time = None, end2_event = None,\n",
    "            advice = False, strata = [], drop = [], **kwargs):\n",
    "    \n",
    "    # Define the keyword variables (function currently accepts up to 10 keyword variables):\n",
    "    var1 = kwargs.get('var1', pd.DataFrame())\n",
    "    var2 = kwargs.get('var2', pd.DataFrame())\n",
    "    var3 = kwargs.get('var3', pd.DataFrame())\n",
    "    var4 = kwargs.get('var4', pd.DataFrame())\n",
    "    var5 = kwargs.get('var5', pd.DataFrame())\n",
    "    var6 = kwargs.get('var6', pd.DataFrame())\n",
    "    var7 = kwargs.get('var7', pd.DataFrame())\n",
    "    var8 = kwargs.get('var8', pd.DataFrame())\n",
    "    var9 = kwargs.get('var9', pd.DataFrame())\n",
    "    var10 = kwargs.get('var10', pd.DataFrame())\n",
    "\n",
    "    var1_name = kwargs.get('var1_name', str)\n",
    "    var2_name = kwargs.get('var2_name', str)\n",
    "    var3_name = kwargs.get('var3_name', str)\n",
    "    var4_name = kwargs.get('var4_name', str)\n",
    "    var5_name = kwargs.get('var5_name', str)\n",
    "    var6_name = kwargs.get('var6_name', str)\n",
    "    var7_name = kwargs.get('var7_name', str)\n",
    "    var8_name = kwargs.get('var8_name', str)\n",
    "    var9_name = kwargs.get('var9_name', str)\n",
    "    var10_name = kwargs.get('var10_name', str)\n",
    "    \n",
    "    dummy_var1 = kwargs.get('dummy_var1', pd.DataFrame())\n",
    "    dummy_var2 = kwargs.get('dummy_var2', pd.DataFrame())\n",
    "    dummy_var3 = kwargs.get('dummy_var3', pd.DataFrame())\n",
    "    dummy_var4 = kwargs.get('dummy_var4', pd.DataFrame())\n",
    "    dummy_var5 = kwargs.get('dummy_var5', pd.DataFrame())\n",
    "    \n",
    "    dummy_var1_name = kwargs.get('dummy_var1_name', tuple())\n",
    "    dummy_var2_name = kwargs.get('dummy_var2_name', tuple())\n",
    "    dummy_var3_name = kwargs.get('dummy_var3_name', tuple())\n",
    "    dummy_var4_name = kwargs.get('dummy_var4_name', tuple())\n",
    "    dummy_var5_name = kwargs.get('dummy_var5_name', tuple())\n",
    "    \n",
    "    # Create a end_times/event dictionary: \n",
    "    end_times_dict = {end1_name : end1_time, \n",
    "                      end2_name : end2_time}\n",
    "    \n",
    "    end_events_dict = {end1_name : end1_event, \n",
    "                       end2_name : end2_event}\n",
    "    \n",
    "    # Create a variables dictionary: \n",
    "    variables = [kwargs[keyword] for keyword in kwargs.keys() if not keyword.endswith('name') if keyword.startswith('var')]\n",
    "    variables_keys = [kwargs[keyword] for keyword in kwargs.keys() if keyword.endswith('name') if keyword.startswith('var')]\n",
    "    variables_dict = dict(zip(variables_keys, variables)) \n",
    "  \n",
    "    # Create a dummy variables dictionary: \n",
    "    dummy_variables = [kwargs[keyword] for keyword in kwargs.keys() if not keyword.endswith('name') if keyword.startswith('dummy')]\n",
    "    dummy_variables_keys = [kwargs[keyword][0] for keyword in kwargs.keys() if keyword.endswith('name') if keyword.startswith('dummy')]\n",
    "    dummy_variables_dict = dict(zip(dummy_variables_keys, dummy_variables))\n",
    "\n",
    "    # Create a dummy ref dictionary: \n",
    "    dummy_ref = [kwargs[keyword][1] for keyword in kwargs.keys() if keyword.endswith('name') if keyword.startswith('dummy')]\n",
    "    dummy_ref_keys = [kwargs[keyword][0] for keyword in kwargs.keys() if keyword.endswith('name') if keyword.startswith('dummy')]\n",
    "    dummy_ref_dict = dict(zip(dummy_ref_keys, dummy_ref))\n",
    "    \n",
    "    # Add dummy variables to variables_dict: \n",
    "    dummy_columns = {}\n",
    "    na = defaultdict(dict)\n",
    "    for var in dummy_variables_dict.keys(): \n",
    "        \n",
    "        if var not in strata: \n",
    "            \n",
    "            # Safe number of NaN for each dummy variable:\n",
    "            for end in [end1_name, end2_name]:\n",
    "                if end:\n",
    "                    na[end][var] = dummy_variables_dict[var].isna().sum()\n",
    "        \n",
    "            # Dropna : \n",
    "            dummy_variables_dict[var] = dummy_variables_dict[var].dropna()\n",
    "        \n",
    "            # Create dummies and drop ref variable: \n",
    "            temporary_df = pd.get_dummies(dummy_variables_dict[var]).drop(dummy_ref_dict[var], axis = 1)\n",
    "        \n",
    "            # Also initialize a list for dummy variable name storing: \n",
    "            dummy_columns[var] = list()\n",
    "\n",
    "            for column in temporary_df.columns: \n",
    "                variables_dict[column] = temporary_df[column]\n",
    "            \n",
    "                # Also store dummy variable names: \n",
    "                dummy_columns[var].append(column)\n",
    "        \n",
    "        else:       \n",
    "            variables_dict[var] = dummy_variables_dict[var]\n",
    "            \n",
    "            \n",
    "    cph_dict = defaultdict(dict)\n",
    "    # For each end, fit a cph model:\n",
    "    for end in [end1_name, end2_name]: \n",
    "        if end: \n",
    "            print('Fitting:', end, '...')\n",
    "        \n",
    "            # For each variable, store non-NaN index values: \n",
    "            ix = {}\n",
    "            for var in variables_dict.keys():\n",
    "                \n",
    "                ix[var] = variables_dict[var].dropna().index.values\n",
    "                \n",
    "                # For each patient, also store number of NaN for plotting: \n",
    "                na[end][var] = variables_dict[var].isna().sum()\n",
    "            \n",
    "            # Reduce dictionary of non-NaN index values to the intersecting set of all variables:  \n",
    "            ix = functools.reduce(np.intersect1d, [ix[var] for var in ix.keys()])\n",
    "            \n",
    "            # For both end_time and end_event, store number of NaN values under ix: \n",
    "            na[end]['time'] = end_times_dict[end].loc[ix].isna().sum()\n",
    "            na[end]['event'] = end_events_dict[end].loc[ix].isna().sum()\n",
    "            \n",
    "            # Intersect the intersecting non-NaN index of all variables with non-NaN index of end_time and end_event: \n",
    "            ix = end_times_dict[end].loc[ix].dropna().index.values\n",
    "            ix = end_events_dict[end].loc[ix].dropna().index.values\n",
    "            \n",
    "            # For each end, store number of patients in the final intersecting ix (i.e. n): \n",
    "            n = len(ix)\n",
    "            \n",
    "            # For each end, store the number of events considering only patients in ix (i.e. the number of events): \n",
    "            events = end_events_dict[end].loc[ix].sum()\n",
    "            print('Number of events:', events)\n",
    "            \n",
    "            # Create a cph_test pd.DataFrame (columns = unique variables and rows = patients): \n",
    "            end_times_dict[end].columns = [end + '_time']\n",
    "            end_events_dict[end].columns = [end + '_event']\n",
    "            cph_test = pd.concat([end_times_dict[end].loc[ix], end_events_dict[end].loc[ix]], axis = 1, join = 'inner', keys = [end + '_time', end + '_event'])\n",
    "            \n",
    "            # Join variables to pd.DataFrame() on intersecting non-NaN index: \n",
    "            for var in variables_dict.keys():\n",
    "            \n",
    "                variables_dict[var].name = var\n",
    "                cph_test = pd.concat([cph_test, variables_dict[var].loc[ix]], axis = 1, join = 'inner') \n",
    "        \n",
    "            # Check whether concatenation worked: \n",
    "            if not (cph_test.index.values == ix).all(): \n",
    "                raise Exception('Concatenated cph_test pd.DataFrame index does not match intersecting set ix')\n",
    "            \n",
    "            if not pd.Series(variables_keys).isin(cph_test.columns).all(): # this one should be the other way around but I leave it for now because it captures when a variable is not a pd.Series\n",
    "                raise Exception('Concatenated cph_test pd.DataFrame does not contain all variables as columns')\n",
    "\n",
    "         \n",
    "            cph = CoxPHFitter()\n",
    "            \n",
    "            if strata:\n",
    "                cph.fit(cph_test, duration_col= end + '_time', event_col= end + '_event', step_size = 0.25, strata = strata)\n",
    "            \n",
    "            else:\n",
    "                cph.fit(cph_test, duration_col= end + '_time', event_col= end + '_event', step_size = 0.25)\n",
    "                \n",
    "       \n",
    "            # Check the assumptions of the cph model: \n",
    "            cph.check_assumptions(cph_test, advice = advice, show_plots = advice)\n",
    "            print('Concordance index:', round(cph.concordance_index_, 3))\n",
    "            \n",
    "            if cph.concordance_index_ < 0.5: \n",
    "                'Concordance is lower than the expected result from random predictions!'\n",
    "            \n",
    "            # Store cph.summary in variable cph_summary for the following operations: \n",
    "            cph_summary = cph.summary \n",
    "            \n",
    "            # Change empty dummy variable NaN values from 0 into '': \n",
    "            for var in dummy_columns.keys():\n",
    "                \n",
    "                for key in dummy_columns[var]:\n",
    "                    na[end][key] = '' # Don't delete the value as the empty string is needed to plot nothing instead of NaN \n",
    "  \n",
    "                # Add dummy ref levels to cph.summary: \n",
    "                where = np.where(cph_summary.index == dummy_columns[var][0])[0][0]\n",
    "                cph_summary = pd.concat([cph_summary.iloc[:where, :], pd.DataFrame({}, index = [var]), cph_summary.iloc[where:, :]])\n",
    "            \n",
    "            # Add stored NaN numbers to cph_summary: \n",
    "            na_concat = pd.DataFrame(pd.Series(na[end]), columns = ['number of NaN'])\n",
    "            cph_summary = cph_summary.join(na_concat, how = 'left')\n",
    "            \n",
    "            if drop:\n",
    "                cph_summary = cph_summary.drop(drop, axis = 0)\n",
    "            \n",
    "            # Store results in end_results_dict:\n",
    "            cph_dict[end]['cph'] = cph_summary # cph fit result \n",
    "            cph_dict[end]['n'] = n # int = number of non-NaN (i.e. complete) patients\n",
    "            cph_dict[end]['events'] = events \n",
    "            cph_dict[end]['na'] = na[end] # dict with keys = variable or time/event and values = number of NaN\n",
    "            cph_dict[end]['ix'] = ix\n",
    "            \n",
    "            print('Finished fitting: ' + end + '!\\n')\n",
    "        \n",
    "    return cph_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_forest_cph(cph_summary, sort_on_pvalues = False, title = '', xlabel = '', fig_adjust = (5, 1), save = '', show = False, nan_str = True):\n",
    "    \n",
    "    cph_summary.index = cph_summary.index.astype(str)\n",
    "    cph_summary = cph_summary.iloc[::-1]\n",
    "    indexLen = np.array([len(index) for index in cph_summary.index]).max()\n",
    "\n",
    "    a, fig, gs = mm.startfig(indexLen + cph_summary['exp(coef) upper 95%'].max() + fig_adjust[0], fig_adjust[1]*len(cph_summary['-log2(p)']))\n",
    "    \n",
    "    if sort_on_pvalues:\n",
    "        cph_summary = cph_summary.sort_values('-log2(p)')\n",
    "    \n",
    "    ylabels_first = []\n",
    "    ylabels_second = []\n",
    "\n",
    "    for hazard, y, state, hazard_lower, hazard_upper, pvalue, nnan, color in zip(np.log10(cph_summary['exp(coef)']), \n",
    "                                                                                 range(0, len(cph_summary['p'])), \n",
    "                                                                                 cph_summary.index, \n",
    "                                                                                 np.log10(cph_summary['exp(coef) lower 95%']), \n",
    "                                                                                 np.log10(cph_summary['exp(coef) upper 95%']),\n",
    "                                                                                 cph_summary['p'], \n",
    "                                                                                 cph_summary['number of NaN'], \n",
    "                                                                                 itertools.cycle(['lightgrey', 'white'])):\n",
    "        \n",
    "        if pvalue < 0.05: \n",
    "            a.scatter(hazard, y, s = 25, c = 'black', zorder = 4, marker = 's')\n",
    "            a.plot([hazard_lower, hazard_upper], [y, y], c = 'black', lw = 1, zorder = 2)\n",
    "        \n",
    "        else:\n",
    "            a.scatter(hazard, y, s = 25, c = 'grey', zorder = 4, marker = 's')\n",
    "            a.plot([hazard_lower, hazard_upper], [y, y], c = 'grey', lw = 1, zorder = 2)\n",
    "    \n",
    "        # 10**hazard + 95% CI str \n",
    "        if np.isnan(pvalue):\n",
    "            \n",
    "            if nan_str: \n",
    "                hazard_str = str(nnan)\n",
    "            \n",
    "            else: \n",
    "                hazard_str = ''\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            if nan_str:\n",
    "                hazard_str = '{:.2f}'.format(round(10**hazard, 2)) + ' (' + '{:.2f}'.format(round(10**hazard_lower, 2)) + '-' + '{:.2f}'.format(round(10**hazard_upper,2)) + ') ' + '    {:.2f}'.format(round(np.log10(pvalue),2)) + ' ' + str(nnan)\n",
    "                \n",
    "            else:\n",
    "                hazard_str = '{:.2f}'.format(round(10**hazard, 2)) + ' (' + '{:.2f}'.format(round(10**hazard_lower, 2)) + '-' + '{:.2f}'.format(round(10**hazard_upper,2)) + ') ' + '    {:.2f}'.format(round(np.log10(pvalue),2))\n",
    "                \n",
    "        ylabels_first.append(state)\n",
    "        ylabels_second.append(hazard_str)\n",
    "\n",
    "        a.axhspan(y - 0.5, y + 0.5, facecolor = color, clip_on = False, zorder = -2)\n",
    "        #a.add_patch(patches.Rectangle((-indexLen/31, y-0.5), abs(-indexLen/31) + cph_summary['exp(coef) upper 95%'].max(), 1, alpha=1, facecolor = color, zorder = -2, clip_on=False))\n",
    "            \n",
    "    a.axvline(np.log10(1), zorder = 0, lw = 1, c = 'grey')\n",
    "            \n",
    "    # set yticks\n",
    "    a.set_yticks(range(0, len(cph_summary['-log2(p)'])))\n",
    "    a.set_yticklabels(ylabels_first)\n",
    "\n",
    "    secay = a.secondary_yaxis('right')\n",
    "    secay.set_yticks(range(0, len(cph_summary['-log2(p)'])))\n",
    "    secay.set_yticklabels(ylabels_second)\n",
    "        \n",
    "    a.tick_params(axis=u'both', which=u'both',length=0)\n",
    "    secay.tick_params(axis=u'both', which=u'both',length=0)\n",
    "\n",
    "    # set spines\n",
    "    a.spines['left'].set_visible(False)\n",
    "    a.spines['right'].set_color('white') # matplotlib secondary yaxis bug: I can't make the right spine invisible anymore\n",
    "    a.spines['top'].set_visible(False)\n",
    "\n",
    "    # set title and xaxis label\n",
    "    a.set_title(title)\n",
    "    a.set_xlabel(xlabel)\n",
    "\n",
    "    a.set_ylim(-0.5, len(cph_summary['-log2(p)'])-0.5)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(save, dpi = 400)\n",
    "    \n",
    "    if show == True: \n",
    "        plt.show()\n",
    "    \n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load TCGA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TCGA sample info from PANCAN atlas: https://gdc.cancer.gov/about-data/publications/pancanatlas\n",
    "TCGA_sample_info = pd.read_excel('./data/tcga/TCGA-CDR-SupplementalTableS1.xlsx', sheet_name=0, header=0, index_col=1)\n",
    "TCGA_survival = pd.read_excel('./data/tcga/TCGA-CDR-SupplementalTableS1.xlsx', sheet_name=2, header=0, index_col=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Stage II', 'Stage IV', 'Stage III', '[Not Available]', 'Stage I',\n",
       "       'Stage X', 'Stage IIB', 'Stage IA', 'Stage IIIA', 'Stage IIA',\n",
       "       'Stage IIIC', 'Stage IB', 'Stage IIIB', '[Discrepancy]',\n",
       "       'Stage IVB', 'Stage IVA', 'Stage IIC', '[Not Applicable]',\n",
       "       '[Unknown]', 'Stage IVC', 'I/II NOS', 'Stage 0', 'IS'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TCGA_sample_info['ajcc_pathologic_tumor_stage'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "stages_map = {\n",
    "    'Stage 0': 'Stage 0 & I', # Merge stage 0 and I \n",
    "    'Stage I': 'Stage 0 & I', \n",
    "    'Stage IA': 'Stage 0 & I', \n",
    "    'Stage IB': 'Stage 0 & I', \n",
    "    'I/II NOS': np.nan, # Ambigious \n",
    "    'Stage II': 'Stage II',\n",
    "    'Stage IIA': 'Stage II', \n",
    "    'Stage IIB': 'Stage II', \n",
    "    'Stage IIC': 'Stage II', \n",
    "    'Stage III': 'Stage III', \n",
    "    'Stage IIIA': 'Stage III', \n",
    "    'Stage IIIB': 'Stage III', \n",
    "    'Stage IIIC': 'Stage III', \n",
    "    'Stage IV': 'Stage IV',\n",
    "    'Stage IVA': 'Stage IV', \n",
    "    'Stage IVB': 'Stage IV', \n",
    "    'Stage IVC': 'Stage IV', \n",
    "    'Stage X': np.nan, \n",
    "    '[Not Available]': np.nan, \n",
    "    '[Not Applicable]': np.nan,\n",
    "    '[Discrepancy]': np.nan,\n",
    "    'IS': np.nan, \n",
    "    '[Unknown]': np.nan \n",
    "}\n",
    "\n",
    "gender_map = {\n",
    "    'MALE': 0, \n",
    "    'FEMALE': 1 \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_sample_info['tumor_stage'] = TCGA_sample_info['ajcc_pathologic_tumor_stage'].map(stages_map)\n",
    "TCGA_sample_info['sex'] = TCGA_sample_info['gender'].map(gender_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(333, 111)\n",
      "(331, 111)\n",
      "Regional_Lymph_Node             160\n",
      "Regional_Skin_or_Soft_Tissue     52\n",
      "-                                43\n",
      "Primary_Disease                  41\n",
      "Distant_Metastases               35\n",
      "Name: REGIONAL_VS_PRIMARY, dtype: int64\n",
      "Regional_Lymph_Node             160\n",
      "Regional_Skin_or_Soft_Tissue     52\n",
      "Primary_Disease                  41\n",
      "Distant_Metastases               35\n",
      "Name: REGIONAL_VS_PRIMARY, dtype: int64\n",
      "[nan 'Primary_Disease' 'Regional_Lymph_Node'\n",
      " 'Regional_Skin_or_Soft_Tissue' 'Distant_Metastases']\n"
     ]
    }
   ],
   "source": [
    "# Load SKCM additional sample info\n",
    "SKCM_info = pd.read_excel('/Users/mariusmessemaker/Documents/Project/mempel/data/tcga/SKCM.xlsx', sheet_name = 3, header = 1, index_col = 0)\n",
    "SKCM_info.index = SKCM_info.index.str[0:12]\n",
    "print(SKCM_info.shape)\n",
    "SKCM_info = SKCM_info.iloc[~SKCM_info.index.duplicated(), :]\n",
    "print(SKCM_info.shape)\n",
    "\n",
    "regional_vs_primary_SKCM_map = {\n",
    "    '-' : np.nan, \n",
    "    'Primary_Disease' : 'Primary_Disease', \n",
    "    'Regional_Lymph_Node' : 'Regional_Lymph_Node', \n",
    "    'Regional_Skin_or_Soft_Tissue' : 'Regional_Skin_or_Soft_Tissue', \n",
    "    'Distant_Metastases': 'Distant_Metastases'\n",
    "}\n",
    "\n",
    "print(SKCM_info['REGIONAL_VS_PRIMARY'].value_counts())\n",
    "SKCM_info['REGIONAL_VS_PRIMARY'] = SKCM_info['REGIONAL_VS_PRIMARY'].map(regional_vs_primary_SKCM_map)\n",
    "print(SKCM_info['REGIONAL_VS_PRIMARY'].value_counts())\n",
    "print(SKCM_info['REGIONAL_VS_PRIMARY'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/pandas/core/frame.py:4304: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    }
   ],
   "source": [
    "# TCGA pancan expression data\n",
    "# gene expression from pancan atlas: https://gdc.cancer.gov/about-data/publications/pancanatlas\n",
    "TCGA_pancan_exp = pd.read_csv('./data/tcga/EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv', sep='\\t', header=0, index_col=0)\n",
    "\n",
    "# sample codes from code table https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes\n",
    "tumor_codes = ['01', '02', '05', '06'] # 06 is also metastatic sites \n",
    "normal_codes = ['10', '11', '12', '13', '14']\n",
    "\n",
    "# divide samples into tumors and normals\n",
    "stype = ['tumor' if x.split('-')[3][0:2] in tumor_codes else 'normal' if x.split('-')[3][0:2] in normal_codes else 'NA' for x in TCGA_pancan_exp.columns.values]\n",
    "\n",
    "sample_type = pd.Series(stype, index = TCGA_pancan_exp.columns.values)\n",
    "\n",
    "#change index to gene names only\n",
    "TCGA_pancan_exp.index = map(lambda x: x.split('|')[0], TCGA_pancan_exp.index.values)\n",
    "\n",
    "# extract tumors only\n",
    "tumor_exp = TCGA_pancan_exp[sample_type[sample_type=='tumor'].index.values]\n",
    "\n",
    "# extract normal samples only\n",
    "normal_exp = TCGA_pancan_exp[sample_type[sample_type=='normal'].index.values]\n",
    "\n",
    "# shorten column names/patient IDs\n",
    "tumor_exp.rename(columns = lambda x: x[0:12], inplace=True)\n",
    "normal_exp.rename(columns = lambda x: x[0:12], inplace=True)\n",
    "\n",
    "# remove duplicate gene expression data for the same sample\n",
    "tumor_exp_unique = tumor_exp.iloc[:, ~tumor_exp.columns.duplicated()].T\n",
    "normal_exp_unique = normal_exp.iloc[:, ~normal_exp.columns.duplicated()].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform expression matrix to dense ranking for DC gene scoring \n",
    "tumor_dense_rank_unique = tumor_exp_unique.transform('rank', method = 'dense', axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Score genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "state_genes = {\n",
    "'CD8B': ['CD8B'],\n",
    "'CCR1': ['CCR1'],\n",
    "'CCR10': ['CCR10'], \n",
    "'CCR2': ['CCR2'], \n",
    "'CCR3': ['CCR3'], \n",
    "'CCR4': ['CCR4'], \n",
    "'CCR5': ['CCR5'], \n",
    "'CCR6':['CCR6'], \n",
    "'CCR7':['CCR7'], \n",
    "'CCR8':['CCR8'], \n",
    "'CCR9':['CCR9'], \n",
    "'CX3CR1':['CX3CR1'],\n",
    "'CXCR1':['CXCR1'], \n",
    "'CXCR2':['CXCR2'],\n",
    "'CXCR3':['CXCR3'],\n",
    "'CXCR4':['CXCR4'],\n",
    "'CXCR5':['CXCR5'],\n",
    "'CXCR6':['CXCR6'], \n",
    "'CXCR7':['CXCR7'],  \n",
    "'XCR1':['XCR1'],\n",
    "'CD19':['CD19'],\n",
    "'CD4':['CD4'],\n",
    "'CXCL16':['CXCL16'],\n",
    "'IL12B':['IL12B'],\n",
    "'NK cells':['NCR1', 'SH2D1B']} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancers = ['BRCA', 'HNSC', 'LUAD', 'LUSC', 'STAD', 'KIRC', 'BLCA', 'LIHC', 'SKCM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA\n",
      "['BRCA'] n= 1095\n",
      "Started scoring: CD8B ...\n",
      "1095\n",
      "Started scoring: CCR1 ...\n",
      "1095\n",
      "Started scoring: CCR10 ...\n",
      "1095\n",
      "Started scoring: CCR2 ...\n",
      "1095\n",
      "Started scoring: CCR3 ...\n",
      "1095\n",
      "Started scoring: CCR4 ...\n",
      "1095\n",
      "Started scoring: CCR5 ...\n",
      "1095\n",
      "Started scoring: CCR6 ...\n",
      "1095\n",
      "Started scoring: CCR7 ...\n",
      "1095\n",
      "Started scoring: CCR8 ...\n",
      "1095\n",
      "Started scoring: CCR9 ...\n",
      "1095\n",
      "Started scoring: CX3CR1 ...\n",
      "1095\n",
      "Started scoring: CXCR1 ...\n",
      "1095\n",
      "Started scoring: CXCR2 ...\n",
      "1095\n",
      "Started scoring: CXCR3 ...\n",
      "1095\n",
      "Started scoring: CXCR4 ...\n",
      "1095\n",
      "Started scoring: CXCR5 ...\n",
      "1095\n",
      "Started scoring: CXCR6 ...\n",
      "1095\n",
      "Started scoring: CXCR7 ...\n",
      "1095\n",
      "Started scoring: XCR1 ...\n",
      "1095\n",
      "Started scoring: CD19 ...\n",
      "1095\n",
      "Started scoring: CD4 ...\n",
      "1095\n",
      "Started scoring: CXCL16 ...\n",
      "1095\n",
      "Started scoring: IL12B ...\n",
      "1095\n",
      "Started scoring: NK cells ...\n",
      "1095\n",
      "HNSC\n",
      "['HNSC'] n= 520\n",
      "Started scoring: CD8B ...\n",
      "520\n",
      "Started scoring: CCR1 ...\n",
      "520\n",
      "Started scoring: CCR10 ...\n",
      "520\n",
      "Started scoring: CCR2 ...\n",
      "520\n",
      "Started scoring: CCR3 ...\n",
      "520\n",
      "Started scoring: CCR4 ...\n",
      "520\n",
      "Started scoring: CCR5 ...\n",
      "520\n",
      "Started scoring: CCR6 ...\n",
      "520\n",
      "Started scoring: CCR7 ...\n",
      "520\n",
      "Started scoring: CCR8 ...\n",
      "520\n",
      "Started scoring: CCR9 ...\n",
      "520\n",
      "Started scoring: CX3CR1 ...\n",
      "520\n",
      "Started scoring: CXCR1 ...\n",
      "520\n",
      "Started scoring: CXCR2 ...\n",
      "520\n",
      "Started scoring: CXCR3 ...\n",
      "520\n",
      "Started scoring: CXCR4 ...\n",
      "520\n",
      "Started scoring: CXCR5 ...\n",
      "520\n",
      "Started scoring: CXCR6 ...\n",
      "520\n",
      "Started scoring: CXCR7 ...\n",
      "520\n",
      "Started scoring: XCR1 ...\n",
      "520\n",
      "Started scoring: CD19 ...\n",
      "520\n",
      "Started scoring: CD4 ...\n",
      "520\n",
      "Started scoring: CXCL16 ...\n",
      "520\n",
      "Started scoring: IL12B ...\n",
      "520\n",
      "Started scoring: NK cells ...\n",
      "520\n",
      "LUAD\n",
      "['LUAD'] n= 515\n",
      "Started scoring: CD8B ...\n",
      "515\n",
      "Started scoring: CCR1 ...\n",
      "515\n",
      "Started scoring: CCR10 ...\n",
      "515\n",
      "Started scoring: CCR2 ...\n",
      "515\n",
      "Started scoring: CCR3 ...\n",
      "515\n",
      "Started scoring: CCR4 ...\n",
      "515\n",
      "Started scoring: CCR5 ...\n",
      "515\n",
      "Started scoring: CCR6 ...\n",
      "515\n",
      "Started scoring: CCR7 ...\n",
      "515\n",
      "Started scoring: CCR8 ...\n",
      "515\n",
      "Started scoring: CCR9 ...\n",
      "515\n",
      "Started scoring: CX3CR1 ...\n",
      "515\n",
      "Started scoring: CXCR1 ...\n",
      "515\n",
      "Started scoring: CXCR2 ...\n",
      "515\n",
      "Started scoring: CXCR3 ...\n",
      "515\n",
      "Started scoring: CXCR4 ...\n",
      "515\n",
      "Started scoring: CXCR5 ...\n",
      "515\n",
      "Started scoring: CXCR6 ...\n",
      "515\n",
      "Started scoring: CXCR7 ...\n",
      "515\n",
      "Started scoring: XCR1 ...\n",
      "515\n",
      "Started scoring: CD19 ...\n",
      "515\n",
      "Started scoring: CD4 ...\n",
      "515\n",
      "Started scoring: CXCL16 ...\n",
      "515\n",
      "Started scoring: IL12B ...\n",
      "515\n",
      "Started scoring: NK cells ...\n",
      "515\n",
      "LUSC\n",
      "['LUSC'] n= 501\n",
      "Started scoring: CD8B ...\n",
      "501\n",
      "Started scoring: CCR1 ...\n",
      "501\n",
      "Started scoring: CCR10 ...\n",
      "501\n",
      "Started scoring: CCR2 ...\n",
      "501\n",
      "Started scoring: CCR3 ...\n",
      "501\n",
      "Started scoring: CCR4 ...\n",
      "501\n",
      "Started scoring: CCR5 ...\n",
      "501\n",
      "Started scoring: CCR6 ...\n",
      "501\n",
      "Started scoring: CCR7 ...\n",
      "501\n",
      "Started scoring: CCR8 ...\n",
      "501\n",
      "Started scoring: CCR9 ...\n",
      "501\n",
      "Started scoring: CX3CR1 ...\n",
      "501\n",
      "Started scoring: CXCR1 ...\n",
      "501\n",
      "Started scoring: CXCR2 ...\n",
      "501\n",
      "Started scoring: CXCR3 ...\n",
      "501\n",
      "Started scoring: CXCR4 ...\n",
      "501\n",
      "Started scoring: CXCR5 ...\n",
      "501\n",
      "Started scoring: CXCR6 ...\n",
      "501\n",
      "Started scoring: CXCR7 ...\n",
      "501\n",
      "Started scoring: XCR1 ...\n",
      "501\n",
      "Started scoring: CD19 ...\n",
      "501\n",
      "Started scoring: CD4 ...\n",
      "501\n",
      "Started scoring: CXCL16 ...\n",
      "501\n",
      "Started scoring: IL12B ...\n",
      "501\n",
      "Started scoring: NK cells ...\n",
      "501\n",
      "STAD\n",
      "['STAD'] n= 415\n",
      "Started scoring: CD8B ...\n",
      "415\n",
      "Started scoring: CCR1 ...\n",
      "415\n",
      "Started scoring: CCR10 ...\n",
      "415\n",
      "Started scoring: CCR2 ...\n",
      "415\n",
      "Started scoring: CCR3 ...\n",
      "415\n",
      "Started scoring: CCR4 ...\n",
      "415\n",
      "Started scoring: CCR5 ...\n",
      "415\n",
      "Started scoring: CCR6 ...\n",
      "415\n",
      "Started scoring: CCR7 ...\n",
      "415\n",
      "Started scoring: CCR8 ...\n",
      "415\n",
      "Started scoring: CCR9 ...\n",
      "415\n",
      "Started scoring: CX3CR1 ...\n",
      "415\n",
      "Started scoring: CXCR1 ...\n",
      "415\n",
      "Started scoring: CXCR2 ...\n",
      "415\n",
      "Started scoring: CXCR3 ...\n",
      "415\n",
      "Started scoring: CXCR4 ...\n",
      "415\n",
      "Started scoring: CXCR5 ...\n",
      "415\n",
      "Started scoring: CXCR6 ...\n",
      "415\n",
      "Started scoring: CXCR7 ...\n",
      "415\n",
      "Started scoring: XCR1 ...\n",
      "415\n",
      "Started scoring: CD19 ...\n",
      "415\n",
      "Started scoring: CD4 ...\n",
      "415\n",
      "Started scoring: CXCL16 ...\n",
      "415\n",
      "Started scoring: IL12B ...\n",
      "415\n",
      "Started scoring: NK cells ...\n",
      "415\n",
      "KIRC\n",
      "['KIRC'] n= 533\n",
      "Started scoring: CD8B ...\n",
      "533\n",
      "Started scoring: CCR1 ...\n",
      "533\n",
      "Started scoring: CCR10 ...\n",
      "533\n",
      "Started scoring: CCR2 ...\n",
      "533\n",
      "Started scoring: CCR3 ...\n",
      "533\n",
      "Started scoring: CCR4 ...\n",
      "533\n",
      "Started scoring: CCR5 ...\n",
      "533\n",
      "Started scoring: CCR6 ...\n",
      "533\n",
      "Started scoring: CCR7 ...\n",
      "533\n",
      "Started scoring: CCR8 ...\n",
      "533\n",
      "Started scoring: CCR9 ...\n",
      "533\n",
      "Started scoring: CX3CR1 ...\n",
      "533\n",
      "Started scoring: CXCR1 ...\n",
      "533\n",
      "Started scoring: CXCR2 ...\n",
      "533\n",
      "Started scoring: CXCR3 ...\n",
      "533\n",
      "Started scoring: CXCR4 ...\n",
      "533\n",
      "Started scoring: CXCR5 ...\n",
      "533\n",
      "Started scoring: CXCR6 ...\n",
      "533\n",
      "Started scoring: CXCR7 ...\n",
      "533\n",
      "Started scoring: XCR1 ...\n",
      "533\n",
      "Started scoring: CD19 ...\n",
      "533\n",
      "Started scoring: CD4 ...\n",
      "533\n",
      "Started scoring: CXCL16 ...\n",
      "533\n",
      "Started scoring: IL12B ...\n",
      "533\n",
      "Started scoring: NK cells ...\n",
      "533\n",
      "BLCA\n",
      "['BLCA'] n= 408\n",
      "Started scoring: CD8B ...\n",
      "408\n",
      "Started scoring: CCR1 ...\n",
      "408\n",
      "Started scoring: CCR10 ...\n",
      "408\n",
      "Started scoring: CCR2 ...\n",
      "408\n",
      "Started scoring: CCR3 ...\n",
      "408\n",
      "Started scoring: CCR4 ...\n",
      "408\n",
      "Started scoring: CCR5 ...\n",
      "408\n",
      "Started scoring: CCR6 ...\n",
      "408\n",
      "Started scoring: CCR7 ...\n",
      "408\n",
      "Started scoring: CCR8 ...\n",
      "408\n",
      "Started scoring: CCR9 ...\n",
      "408\n",
      "Started scoring: CX3CR1 ...\n",
      "408\n",
      "Started scoring: CXCR1 ...\n",
      "408\n",
      "Started scoring: CXCR2 ...\n",
      "408\n",
      "Started scoring: CXCR3 ...\n",
      "408\n",
      "Started scoring: CXCR4 ...\n",
      "408\n",
      "Started scoring: CXCR5 ...\n",
      "408\n",
      "Started scoring: CXCR6 ...\n",
      "408\n",
      "Started scoring: CXCR7 ...\n",
      "408\n",
      "Started scoring: XCR1 ...\n",
      "408\n",
      "Started scoring: CD19 ...\n",
      "408\n",
      "Started scoring: CD4 ...\n",
      "408\n",
      "Started scoring: CXCL16 ...\n",
      "408\n",
      "Started scoring: IL12B ...\n",
      "408\n",
      "Started scoring: NK cells ...\n",
      "408\n",
      "LIHC\n",
      "['LIHC'] n= 371\n",
      "Started scoring: CD8B ...\n",
      "371\n",
      "Started scoring: CCR1 ...\n",
      "371\n",
      "Started scoring: CCR10 ...\n",
      "371\n",
      "Started scoring: CCR2 ...\n",
      "371\n",
      "Started scoring: CCR3 ...\n",
      "371\n",
      "Started scoring: CCR4 ...\n",
      "371\n",
      "Started scoring: CCR5 ...\n",
      "371\n",
      "Started scoring: CCR6 ...\n",
      "371\n",
      "Started scoring: CCR7 ...\n",
      "371\n",
      "Started scoring: CCR8 ...\n",
      "371\n",
      "Started scoring: CCR9 ...\n",
      "371\n",
      "Started scoring: CX3CR1 ...\n",
      "371\n",
      "Started scoring: CXCR1 ...\n",
      "371\n",
      "Started scoring: CXCR2 ...\n",
      "371\n",
      "Started scoring: CXCR3 ...\n",
      "371\n",
      "Started scoring: CXCR4 ...\n",
      "371\n",
      "Started scoring: CXCR5 ...\n",
      "371\n",
      "Started scoring: CXCR6 ...\n",
      "371\n",
      "Started scoring: CXCR7 ...\n",
      "371\n",
      "Started scoring: XCR1 ...\n",
      "371\n",
      "Started scoring: CD19 ...\n",
      "371\n",
      "Started scoring: CD4 ...\n",
      "371\n",
      "Started scoring: CXCL16 ...\n",
      "371\n",
      "Started scoring: IL12B ...\n",
      "371\n",
      "Started scoring: NK cells ...\n",
      "371\n",
      "SKCM\n",
      "['SKCM'] n= 469\n",
      "Started scoring: CD8B ...\n",
      "469\n",
      "Started scoring: CCR1 ...\n",
      "469\n",
      "Started scoring: CCR10 ...\n",
      "469\n",
      "Started scoring: CCR2 ...\n",
      "469\n",
      "Started scoring: CCR3 ...\n",
      "469\n",
      "Started scoring: CCR4 ...\n",
      "469\n",
      "Started scoring: CCR5 ...\n",
      "469\n",
      "Started scoring: CCR6 ...\n",
      "469\n",
      "Started scoring: CCR7 ...\n",
      "469\n",
      "Started scoring: CCR8 ...\n",
      "469\n",
      "Started scoring: CCR9 ...\n",
      "469\n",
      "Started scoring: CX3CR1 ...\n",
      "469\n",
      "Started scoring: CXCR1 ...\n",
      "469\n",
      "Started scoring: CXCR2 ...\n",
      "469\n",
      "Started scoring: CXCR3 ...\n",
      "469\n",
      "Started scoring: CXCR4 ...\n",
      "469\n",
      "Started scoring: CXCR5 ...\n",
      "469\n",
      "Started scoring: CXCR6 ...\n",
      "469\n",
      "Started scoring: CXCR7 ...\n",
      "469\n",
      "Started scoring: XCR1 ...\n",
      "469\n",
      "Started scoring: CD19 ...\n",
      "469\n",
      "Started scoring: CD4 ...\n",
      "469\n",
      "Started scoring: CXCL16 ...\n",
      "469\n",
      "Started scoring: IL12B ...\n",
      "469\n",
      "Started scoring: NK cells ...\n",
      "469\n"
     ]
    }
   ],
   "source": [
    "log_zscore = True\n",
    "\n",
    "state_scores = defaultdict(dict)\n",
    "for cancer in cancers:\n",
    "    print(cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer + '/distribution/')\n",
    "    \n",
    "    # Exclude samples that are metastatic (or lymph-node metastatic in SKCM)\n",
    "    #if cancer in exclude_dict.keys():\n",
    "     #   ix0 = TCGA_sample_info[~TCGA_sample_info.index.isin(exclude_dict[cancer])].index.values        \n",
    "      #  ix1 = TCGA_sample_info[TCGA_sample_info['type']==cancer].index.values \n",
    "       #  ix = np.intersect1d(ix0, ix1)\n",
    "    \n",
    "    #elif cancer == 'SKCM': \n",
    "     #   ix = SKCM_info[~SKCM_info.index.isin(SKCM_ix)].index.values \n",
    "        \n",
    "    ix = TCGA_sample_info[TCGA_sample_info['type']==cancer].index.values\n",
    "    \n",
    "    # Check whether indexing worked: \n",
    "    print(TCGA_sample_info.loc[ix]['type'].unique(), 'n=', len([sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index]))\n",
    "    \n",
    "    for state in state_genes.keys():\n",
    "        print('Started scoring:', state, '...')\n",
    "        \n",
    "        # Check whether genes are detected \n",
    "        genes_first = [gene for gene in state_genes[state] if gene in tumor_exp_unique.columns]\n",
    "        genes = []\n",
    "        \n",
    "        for gene in genes_first:\n",
    "            \n",
    "            # Check whether genes are NaN although they are detected \n",
    "            if tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index], gene].isna().sum() < 1:\n",
    "                genes.append(gene)\n",
    "                \n",
    "            else:\n",
    "                # If a gene is not included in the columns or NaN: \n",
    "                print('Not detected:', gene)\n",
    "        \n",
    "        # Plot distribution before gene scoring \n",
    "        sns.displot(tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index], genes].mean(1))\n",
    "        plt.title('Distribution ' + state + ' before scoring')\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/distribution/' + state + '_before.pdf', dpi = 300)\n",
    "        plt.close()\n",
    "        \n",
    "        # Score genes \n",
    "        if log_zscore:\n",
    "            # EdgeR pseudo count calculation https://bioinformatics.stackexchange.com/questions/4985/how-is-prior-count-used-by-edgers-cpm\n",
    "            pseudo = (tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index]].sum(1)/tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index]].sum(1).mean())*2*0.25\n",
    "            \n",
    "            if not (pseudo.index == tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index]].index).all():\n",
    "                raise Exception('pseudo index not the same as expression matrix index')\n",
    "            \n",
    "            if len(genes) <= 1:\n",
    "                state_scores[cancer][state] = np.log2(tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index], \n",
    "                                                        genes].add(pseudo, axis = 0)).apply(zscore, axis = 0)[genes[0]]\n",
    "            else:\n",
    "                state_scores[cancer][state] = np.log2(tumor_exp_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_exp_unique.index], \n",
    "                                                      genes].add(pseudo, axis = 0)).apply(zscore, axis = 0).mean(1)\n",
    "                                            \n",
    "        else: \n",
    "            state_scores[cancer][state] = tumor_dense_rank_unique.loc[[sample for sample in TCGA_sample_info.loc[ix].index.values if sample in tumor_dense_rank_unique.index], \n",
    "                                                                       genes].sum(1)/len(genes)\n",
    "            state_scores[cancer][state] = state_scores[cancer][state]/state_scores[cancer][state].max()\n",
    "            \n",
    "        print(len(state_scores[cancer][state]))\n",
    "        \n",
    "        # Plot distribution of scored genes \n",
    "        sns.displot(state_scores[cancer][state])\n",
    "        plt.title('Distribution ' + state + ' after scoring')\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/distribution/' + state + '_after.pdf', dpi = 300)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continious association of scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CD8B', 'CXCL16', 'CXCR6', 'CXCR3', 'CXCR2', 'CCR5', 'CCR10', 'CCR8', 'CD19', 'CD4', 'NK cells', 'IL12B'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "HNSC\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "LUAD\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "LUSC\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "STAD\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "KIRC\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "BLCA\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "LIHC\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n",
      "SKCM\n",
      "Comparing: CD8B_with_CXCL16\n",
      "Comparing: CD8B_with_CXCR6\n",
      "Comparing: CD8B_with_CXCR3\n",
      "Comparing: CD8B_with_CXCR2\n",
      "Comparing: CD8B_with_CCR5\n",
      "Comparing: CD8B_with_CCR10\n",
      "Comparing: CD8B_with_CCR8\n",
      "Comparing: CD8B_with_CD19\n",
      "Comparing: CD8B_with_CD4\n",
      "Comparing: CD8B_with_NK cells\n",
      "Comparing: CD8B_with_IL12B\n",
      "Comparing: CXCL16_with_CXCR6\n",
      "Comparing: CXCL16_with_CXCR3\n",
      "Comparing: CXCL16_with_CXCR2\n",
      "Comparing: CXCL16_with_CCR5\n",
      "Comparing: CXCL16_with_CCR10\n",
      "Comparing: CXCL16_with_CCR8\n",
      "Comparing: CXCL16_with_CD19\n",
      "Comparing: CXCL16_with_CD4\n",
      "Comparing: CXCL16_with_NK cells\n",
      "Comparing: CXCL16_with_IL12B\n",
      "Comparing: CXCR6_with_CXCR3\n",
      "Comparing: CXCR6_with_CXCR2\n",
      "Comparing: CXCR6_with_CCR5\n",
      "Comparing: CXCR6_with_CCR10\n",
      "Comparing: CXCR6_with_CCR8\n",
      "Comparing: CXCR6_with_CD19\n",
      "Comparing: CXCR6_with_CD4\n",
      "Comparing: CXCR6_with_NK cells\n",
      "Comparing: CXCR6_with_IL12B\n",
      "Comparing: CXCR3_with_CXCR2\n",
      "Comparing: CXCR3_with_CCR5\n",
      "Comparing: CXCR3_with_CCR10\n",
      "Comparing: CXCR3_with_CCR8\n",
      "Comparing: CXCR3_with_CD19\n",
      "Comparing: CXCR3_with_CD4\n",
      "Comparing: CXCR3_with_NK cells\n",
      "Comparing: CXCR3_with_IL12B\n",
      "Comparing: CXCR2_with_CCR5\n",
      "Comparing: CXCR2_with_CCR10\n",
      "Comparing: CXCR2_with_CCR8\n",
      "Comparing: CXCR2_with_CD19\n",
      "Comparing: CXCR2_with_CD4\n",
      "Comparing: CXCR2_with_NK cells\n",
      "Comparing: CXCR2_with_IL12B\n",
      "Comparing: CCR5_with_CCR10\n",
      "Comparing: CCR5_with_CCR8\n",
      "Comparing: CCR5_with_CD19\n",
      "Comparing: CCR5_with_CD4\n",
      "Comparing: CCR5_with_NK cells\n",
      "Comparing: CCR5_with_IL12B\n",
      "Comparing: CCR10_with_CCR8\n",
      "Comparing: CCR10_with_CD19\n",
      "Comparing: CCR10_with_CD4\n",
      "Comparing: CCR10_with_NK cells\n",
      "Comparing: CCR10_with_IL12B\n",
      "Comparing: CCR8_with_CD19\n",
      "Comparing: CCR8_with_CD4\n",
      "Comparing: CCR8_with_NK cells\n",
      "Comparing: CCR8_with_IL12B\n",
      "Comparing: CD19_with_CD4\n",
      "Comparing: CD19_with_NK cells\n",
      "Comparing: CD19_with_IL12B\n",
      "Comparing: CD4_with_NK cells\n",
      "Comparing: CD4_with_IL12B\n",
      "Comparing: NK cells_with_IL12B\n"
     ]
    }
   ],
   "source": [
    "for cancer in cancers: \n",
    "    print(cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer + '/scatter/')\n",
    "    \n",
    "    # Compare all combinations: \n",
    "    for state1, state2 in itertools.combinations(states, 2):\n",
    "        name = state1 + '_with_' + state2\n",
    "        print('Comparing:', name)\n",
    "        \n",
    "        # Plot scatterplot that shows relation of the continious variables: \n",
    "        a, fig, gs = mm.startfig(8, 8) \n",
    "        \n",
    "        if not (state_scores[cancer][state1].index == state_scores[cancer][state2].index).all():\n",
    "            raise Excpetion('Indexes are not the same')\n",
    "        \n",
    "        a.scatter(state_scores[cancer][state1], state_scores[cancer][state2], c = 'black', s = 5)\n",
    "        a.set_xlabel(state1 + ' score')\n",
    "        a.set_ylabel(state2 + ' score')\n",
    "        \n",
    "        if spearmanr(state_scores[cancer][state1], state_scores[cancer][state2])[1] < 0.001:\n",
    "            a.set_title(cancer + ' r='+str(round(spearmanr(state_scores[cancer][state1], state_scores[cancer][state2])[0],2)) + ' p < 0.001')\n",
    "        else:\n",
    "            a.set_title(cancer + ' r='+str(round(spearmanr(state_scores[cancer][state1], state_scores[cancer][state2])[0],2)) + ' p=' + str(round(spearmanr(state_scores[cancer][state1], state_scores[cancer][state2])[1],3)))\n",
    "            \n",
    "        plt.locator_params(nbins=10)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/scatter/' + name + '.pdf')\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Association of scores with tumor stage "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CD8B', 'CXCL16', 'CXCR6', 'CXCR3', 'CXCR2', 'CCR5', 'CCR10', 'CD4', 'NK cells', 'IL12B'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA\n",
      "CD8B\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 5.5% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CXCL16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 13.2% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CXCR6\n",
      "CXCR3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 7.4% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CXCR2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 19.5% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCR5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 8.5% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCR10\n",
      "significant!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 9.2% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD4\n",
      "NK cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 12.6% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IL12B\n",
      "HNSC\n",
      "CD8B\n",
      "CXCL16\n",
      "CXCR6\n",
      "significant!\n",
      "CXCR3\n",
      "significant!\n",
      "CXCR2\n",
      "significant!\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "CD4\n",
      "NK cells\n",
      "significant!\n",
      "IL12B\n",
      "LUAD\n",
      "CD8B\n",
      "significant!\n",
      "CXCL16\n",
      "CXCR6\n",
      "significant!\n",
      "CXCR3\n",
      "CXCR2\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "significant!\n",
      "CD4\n",
      "significant!\n",
      "NK cells\n",
      "IL12B\n",
      "significant!\n",
      "LUSC\n",
      "CD8B\n",
      "CXCL16\n",
      "CXCR6\n",
      "CXCR3\n",
      "CXCR2\n",
      "CCR5\n",
      "CCR10\n",
      "CD4\n",
      "NK cells\n",
      "IL12B\n",
      "STAD\n",
      "CD8B\n",
      "CXCL16\n",
      "CXCR6\n",
      "CXCR3\n",
      "significant!\n",
      "CXCR2\n",
      "significant!\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "significant!\n",
      "CD4\n",
      "significant!\n",
      "NK cells\n",
      "IL12B\n",
      "KIRC\n",
      "CD8B\n",
      "significant!\n",
      "CXCL16\n",
      "CXCR6\n",
      "significant!\n",
      "CXCR3\n",
      "significant!\n",
      "CXCR2\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "CD4\n",
      "NK cells\n",
      "IL12B\n",
      "BLCA\n",
      "CD8B\n",
      "CXCL16\n",
      "CXCR6\n",
      "CXCR3\n",
      "CXCR2\n",
      "CCR5\n",
      "CCR10\n",
      "CD4\n",
      "NK cells\n",
      "IL12B\n",
      "LIHC\n",
      "CD8B\n",
      "CXCL16\n",
      "CXCR6\n",
      "CXCR3\n",
      "CXCR2\n",
      "CCR5\n",
      "CCR10\n",
      "CD4\n",
      "NK cells\n",
      "IL12B\n",
      "SKCM\n",
      "CD8B\n",
      "significant!\n",
      "CXCL16\n",
      "CXCR6\n",
      "significant!\n",
      "CXCR3\n",
      "significant!\n",
      "CXCR2\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "CD4\n",
      "significant!\n",
      "NK cells\n",
      "significant!\n",
      "IL12B\n"
     ]
    }
   ],
   "source": [
    "for cancer in cancers: \n",
    "    print(cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer + '/tumor_stage_boxplot/')\n",
    "    \n",
    "    for state in states: \n",
    "        print(state)\n",
    "\n",
    "        score_stage_df = pd.DataFrame(state_scores[cancer][state], columns = [state]).join(TCGA_sample_info['tumor_stage']).dropna()\n",
    "\n",
    "        kruskal = stats.kruskal(score_stage_df[score_stage_df['tumor_stage'] == 'Stage 0 & I'][state],\n",
    "                              score_stage_df[score_stage_df['tumor_stage'] == 'Stage II'][state],\n",
    "                              score_stage_df[score_stage_df['tumor_stage'] == 'Stage III'][state], \n",
    "                              score_stage_df[score_stage_df['tumor_stage'] == 'Stage IV'][state])\n",
    "        \n",
    "        if kruskal[1] < 0.05: \n",
    "            print('significant!')\n",
    "        \n",
    "        order = ['Stage 0 & I', 'Stage II', 'Stage III', 'Stage IV']\n",
    "        ax = sns.swarmplot(x = 'tumor_stage', y = state, data = score_stage_df, color = 'black', zorder = 0, s = 3, order = order)\n",
    "        ax = sns.boxplot(x = 'tumor_stage', y = state, data = score_stage_df, color = 'white', zorder = 4, order = order)\n",
    "        \n",
    "        ax.set_xticklabels(['Stage 0 & I (n=' + str(len(score_stage_df[score_stage_df['tumor_stage'] == 'Stage 0 & I'])) + ')',\n",
    "                            'Stage II (n=' + str(len(score_stage_df[score_stage_df['tumor_stage'] == 'Stage II'])) + ')',\n",
    "                            'Stage III (n=' + str(len(score_stage_df[score_stage_df['tumor_stage'] == 'Stage III'])) + ')',\n",
    "                            'Stage IV (n=' + str(len(score_stage_df[score_stage_df['tumor_stage'] == 'Stage IV'])) + ')'])\n",
    "        \n",
    "        plt.xlabel('')\n",
    "        \n",
    "        if kruskal[1] < 0.001:\n",
    "            ax.set_title(cancer + ' H=' + str(round(kruskal[0], 2)) + ' p < 0.001')\n",
    "        else: \n",
    "            ax.set_title(cancer + ' H=' + str(round(kruskal[0], 2)) + ' p=' + str(round(kruskal[1], 2)))\n",
    "            \n",
    "        for patch in ax.artists:\n",
    "            r, g, b, a = patch.get_facecolor()\n",
    "            patch.set_facecolor((r, g, b, 0))\n",
    "        \n",
    "        plt.ylabel(state + ' score')\n",
    "        plt.tight_layout()\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/' + 'tumor_stage_boxplot/' + state + '.pdf')\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Association scores with lymphocyte score in SKCM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CD8B', 'CXCL16', 'CXCR6', 'CXCR3', 'CXCR2', 'CCR5', 'CCR10', 'CD4', 'NK cells', 'IL12B'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "lymphocyte_score_map = {\n",
    "    0: 'Lc score 0 & 2', \n",
    "    2: 'Lc score 0 & 2', \n",
    "    3: 'Lc score 3 & 4', \n",
    "    4: 'Lc score 3 & 4',\n",
    "    5: 'Lc score 5 & 6',\n",
    "    6: 'Lc score 5 & 6', \n",
    "    '-': np.nan\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "SKCM_info['Lc_score'] = SKCM_info['LYMPHOCYTE.SCORE'].map(lymphocyte_score_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD8B\n",
      "significant!\n",
      "CXCL16\n",
      "significant!\n",
      "CXCR6\n",
      "significant!\n",
      "CXCR3\n",
      "significant!\n",
      "CXCR2\n",
      "significant!\n",
      "CCR5\n",
      "significant!\n",
      "CCR10\n",
      "CD4\n",
      "significant!\n",
      "NK cells\n",
      "significant!\n",
      "IL12B\n",
      "significant!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mariusmessemaker/miniconda3/envs/py36-sc/lib/python3.6/site-packages/seaborn/categorical.py:1296: UserWarning: 12.0% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "cancer = 'SKCM'\n",
    "os.mkdir('./outputs/survival/' + cancer + '/lc_score_boxplot/')\n",
    "for state in states: \n",
    "    print(state)\n",
    "\n",
    "    score_stage_df = pd.DataFrame(state_scores[cancer][state], columns = [state]).join(SKCM_info['Lc_score']).dropna()\n",
    "\n",
    "    kruskal = stats.kruskal(score_stage_df[score_stage_df['Lc_score'] == 'Lc score 0 & 2'][state],\n",
    "                          score_stage_df[score_stage_df['Lc_score'] == 'Lc score 3 & 4'][state],\n",
    "                          score_stage_df[score_stage_df['Lc_score'] == 'Lc score 5 & 6'][state])\n",
    "  \n",
    "        \n",
    "    if kruskal[1] < 0.05: \n",
    "        print('significant!')\n",
    "        \n",
    "    order = ['Lc score 0 & 2', 'Lc score 3 & 4', 'Lc score 5 & 6']\n",
    "    ax = sns.swarmplot(x = 'Lc_score', y = state, data = score_stage_df, color = 'black', zorder = 0, s = 6, order = order)\n",
    "    ax = sns.boxplot(x = 'Lc_score', y = state, data = score_stage_df, color = 'white', zorder = 4, order = order)\n",
    "    \n",
    "    ax.set_xticklabels(['Lc score 0 & 2 (n=' + str(len(score_stage_df[score_stage_df['Lc_score'] == 'Lc score 0 & 2'])) + ')',\n",
    "                        'Lc score 3 & 4 (n=' + str(len(score_stage_df[score_stage_df['Lc_score'] == 'Lc score 3 & 4'])) + ')',\n",
    "                        'Lc score 5 & 6 (n=' + str(len(score_stage_df[score_stage_df['Lc_score'] == 'Lc score 5 & 6'])) + ')'])\n",
    "    plt.xlabel('')\n",
    "               \n",
    "    if kruskal[1] < 0.001:\n",
    "        ax.set_title(cancer + ' H=' + str(round(kruskal[0], 2)) + ' p < 0.001')\n",
    "    else: \n",
    "        ax.set_title(cancer + ' H=' + str(round(kruskal[0], 2)) + ' p=' + str(round(kruskal[1], 3)))\n",
    "            \n",
    "    for patch in ax.artists:\n",
    "        r, g, b, a = patch.get_facecolor()\n",
    "        patch.set_facecolor((r, g, b, 0))\n",
    "        \n",
    "    plt.ylabel(state + ' score')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('./outputs/survival/' + cancer + '/' + 'lc_score_boxplot/' + state + '.pdf')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Univariate CPH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CCR1', 'CCR10', 'CCR2', 'CCR3', 'CCR4', 'CCR5', 'CCR6','CCR7', 'CCR8', \n",
    "          'CCR9', 'CX3CR1', 'CXCR1', 'CXCR2', 'CXCR3', 'CXCR4', 'CXCR5', 'CXCR6', 'CXCR7', 'XCR1'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.492\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.495\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.5\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.518\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.559\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.559\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 1094 total...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR6</th>\n",
       "      <th>km</th>\n",
       "      <td>6.62</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>4.92</td>\n",
       "      <td>0.03</td>\n",
       "      <td>5.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR6 & km &        6.622370 &  0.010071 &  6.633711 \\\\\n",
       "     & rank &        4.918407 &  0.026572 &  5.233949 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 1094 total observations, 943 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic    p  -log2(p)\n",
       "CCR6 km              6.62 0.01      6.63\n",
       "     rank            4.92 0.03      5.23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR6' failed the non-proportional test: p-value is 0.0101.\n",
      "\n",
      "Concordance index: 0.56\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 1094 total...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR6</th>\n",
       "      <th>km</th>\n",
       "      <td>9.93</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>6.90</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR6 & km &        9.933718 &  0.001623 &  9.267313 \\\\\n",
       "     & rank &        6.904697 &  0.008597 &  6.861958 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 1094 total observations, 890 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic      p  -log2(p)\n",
       "CCR6 km              9.93 <0.005      9.27\n",
       "     rank            6.90   0.01      6.86"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR6' failed the non-proportional test: p-value is 0.0016.\n",
      "\n",
      "Concordance index: 0.578\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.501\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.518\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.546\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.572\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.492\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.49\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.564\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.501\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 151.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.583\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 204.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.565\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "HNSC\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.548\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.561\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.534\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.587\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.585\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.559\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.591\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.58\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.606\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.588\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.518\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.509\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 519 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCR1</th>\n",
       "      <th>km</th>\n",
       "      <td>8.37</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>8.24</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>7.93</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "      &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCR1 & km &        8.365162 &  0.003825 &  8.030390 \\\\\n",
       "      & rank &        8.237093 &  0.004104 &  7.928654 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 519 total observations, 299 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic      p  -log2(p)\n",
       "CXCR1 km              8.37 <0.005      8.03\n",
       "      rank            8.24 <0.005      7.93"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCR1' failed the non-proportional test: p-value is 0.0038.\n",
      "\n",
      "Concordance index: 0.538\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 519 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCR1</th>\n",
       "      <th>km</th>\n",
       "      <td>13.58</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>12.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>14.64</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>12.91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "      &      &  test\\_statistic &         p &   -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCR1 & km &       13.578038 &  0.000229 &  12.093327 \\\\\n",
       "      & rank &       14.639631 &  0.000130 &  12.907551 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 519 total observations, 248 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic      p  -log2(p)\n",
       "CXCR1 km             13.58 <0.005     12.09\n",
       "      rank           14.64 <0.005     12.91"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCR1' failed the non-proportional test: p-value is 0.0001.\n",
      "\n",
      "Concordance index: 0.531\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.505\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.5\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.567\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.548\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.546\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.6\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.584\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.501\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 220.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.572\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 271.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LUAD\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.614\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR2</th>\n",
       "      <th>km</th>\n",
       "      <td>7.47</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>7.34</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7.21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR2 & km &        7.473248 &  0.006262 &  7.319104 \\\\\n",
       "     & rank &        7.337700 &  0.006752 &  7.210401 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic    p  -log2(p)\n",
       "CCR2 km              7.47 0.01      7.32\n",
       "     rank            7.34 0.01      7.21"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR2' failed the non-proportional test: p-value is 0.0063.\n",
      "\n",
      "Concordance index: 0.584\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.529\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.586\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.567\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.608\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR6</th>\n",
       "      <th>km</th>\n",
       "      <td>13.06</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>11.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>14.28</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>12.63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &   -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR6 & km &       13.063761 &  0.000301 &  11.697641 \\\\\n",
       "     & rank &       14.281754 &  0.000157 &  12.633432 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic      p  -log2(p)\n",
       "CCR6 km             13.06 <0.005     11.70\n",
       "     rank           14.28 <0.005     12.63"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR6' failed the non-proportional test: p-value is 0.0002.\n",
      "\n",
      "Concordance index: 0.583\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.585\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR7</th>\n",
       "      <th>km</th>\n",
       "      <td>9.22</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>10.37</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR7 & km &        9.220748 &  0.002393 &  8.707038 \\\\\n",
       "     & rank &       10.372854 &  0.001279 &  9.610975 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic      p  -log2(p)\n",
       "CCR7 km              9.22 <0.005      8.71\n",
       "     rank           10.37 <0.005      9.61"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR7' failed the non-proportional test: p-value is 0.0013.\n",
      "\n",
      "Concordance index: 0.563\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.533\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.561\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR9</th>\n",
       "      <th>km</th>\n",
       "      <td>6.89</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>7.22</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "     &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR9 & km &        6.891848 &  0.008659 &  6.851589 \\\\\n",
       "     & rank &        7.218563 &  0.007215 &  7.114716 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic    p  -log2(p)\n",
       "CCR9 km              6.89 0.01      6.85\n",
       "     rank            7.22 0.01      7.11"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR9' failed the non-proportional test: p-value is 0.0072.\n",
      "\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CX3CR1</th>\n",
       "      <th>km</th>\n",
       "      <td>9.41</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>10.02</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "       &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CX3CR1 & km &        9.405142 &  0.002164 &  8.852232 \\\\\n",
       "       & rank &       10.018261 &  0.001550 &  9.333555 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 323 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "             test_statistic      p  -log2(p)\n",
       "CX3CR1 km              9.41 <0.005      8.85\n",
       "       rank           10.02 <0.005      9.33"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CX3CR1' failed the non-proportional test: p-value is 0.0015.\n",
      "\n",
      "Concordance index: 0.605\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CX3CR1</th>\n",
       "      <th>km</th>\n",
       "      <td>12.56</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>11.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>12.84</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>11.53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "       &      &  test\\_statistic &         p &   -log2(p) \\\\\n",
       "\\midrule\n",
       "CX3CR1 & km &       12.561468 &  0.000394 &  11.310315 \\\\\n",
       "       & rank &       12.842550 &  0.000339 &  11.527170 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "             test_statistic      p  -log2(p)\n",
       "CX3CR1 km             12.56 <0.005     11.31\n",
       "       rank           12.84 <0.005     11.53"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CX3CR1' failed the non-proportional test: p-value is 0.0003.\n",
      "\n",
      "Concordance index: 0.584\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.496\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.491\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.546\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.567\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.584\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 506 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCR5</th>\n",
       "      <th>km</th>\n",
       "      <td>8.94</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>8.65</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "      &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCR5 & km &        8.937277 &  0.002794 &  8.483406 \\\\\n",
       "      & rank &        8.654647 &  0.003262 &  8.259902 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 506 total observations, 247 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic      p  -log2(p)\n",
       "CXCR5 km              8.94 <0.005      8.48\n",
       "      rank            8.65 <0.005      8.26"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCR5' failed the non-proportional test: p-value is 0.0028.\n",
      "\n",
      "Concordance index: 0.568\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.565\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 183.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.595\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 259.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.571\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 9, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LUSC\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.531\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.511\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.502\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.506\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.506\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.498\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.487\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.512\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.499\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 496 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CX3CR1</th>\n",
       "      <th>km</th>\n",
       "      <td>6.46</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>6.32</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "       &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CX3CR1 & km &        6.458119 &  0.011045 &  6.500509 \\\\\n",
       "       & rank &        6.323668 &  0.011914 &  6.391235 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 496 total observations, 244 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "             test_statistic    p  -log2(p)\n",
       "CX3CR1 km              6.46 0.01      6.50\n",
       "       rank            6.32 0.01      6.39"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CX3CR1' failed the non-proportional test: p-value is 0.0110.\n",
      "\n",
      "Concordance index: 0.493\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.513\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.496\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.511\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.522\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.509\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.496\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.505\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.511\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.497\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 212.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 252.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 5, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "STAD\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.563\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.561\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.543\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.567\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.512\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.484\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.506\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.531\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.531\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.509\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.503\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.587\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.569\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.589\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 157.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 196.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 4, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "KIRC\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.543\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.582\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.563\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.509\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.565\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.507\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.563\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.566\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.543\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.534\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.555\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.571\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.595\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.588\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.555\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 175.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.6\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 222.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.577\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 2, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "BLCA\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.515\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.518\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.548\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.523\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.518\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.503\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.515\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.497\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 408 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCR2</th>\n",
       "      <th>km</th>\n",
       "      <td>6.64</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>7.16</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7.07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "      &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCR2 & km &        6.644845 &  0.009944 &  6.651912 \\\\\n",
       "      & rank &        7.159756 &  0.007456 &  7.067434 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 408 total observations, 179 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic    p  -log2(p)\n",
       "CXCR2 km              6.64 0.01      6.65\n",
       "      rank            7.16 0.01      7.07"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCR2' failed the non-proportional test: p-value is 0.0075.\n",
      "\n",
      "Concordance index: 0.532\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.533\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.494\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.501\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.522\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.534\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 178.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.496\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 229.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.504\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 0, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LIHC\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.496\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.538\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.571\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.591\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.573\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.602\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.583\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.529\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.497\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.498\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.574\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.499\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.511\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.575\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.569\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.583\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.586\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.515\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 130.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 225.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.581\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 1, 'event': 0}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SKCM\n",
      "CCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.61\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR10\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.515\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.62\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.593\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.577\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.623\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.583\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR8\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.612\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.543\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CCR9\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.576\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CX3CR1\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR2\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR3\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.619\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.557\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR4\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.569\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR5\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.544\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR6\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.624\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "CXCR7\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "XCR1\n",
      "Fitting: OS ...\n",
      "Number of events: 213.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.576\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 322.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "{'XCR1': 0, 'time': 15, 'event': 7}\n",
      "PFS\n",
      "{'XCR1': 0, 'time': 14, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for cancer in cancers:\n",
    "    print(cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer + '/cph/')\n",
    "    \n",
    "    cph_concat = {}\n",
    "    cph_concat['OS'] = pd.DataFrame()\n",
    "    cph_concat['PFS'] = pd.DataFrame()\n",
    "    \n",
    "    for state in states: \n",
    "        print(state)\n",
    "        ix = state_scores[cancer][state].index.values\n",
    "        \n",
    "        cph_dict = fit_cph(end1_name = 'OS', end1_time = TCGA_sample_info[TCGA_sample_info['type']==cancer].loc[ix,'OS.time'], end1_event = TCGA_sample_info[TCGA_sample_info['type']==cancer].loc[ix, 'OS'], \n",
    "                           end2_name = 'PFS', end2_time = TCGA_survival[TCGA_survival['type']==cancer].loc[ix, 'PFS.time'], end2_event = TCGA_survival[TCGA_survival['type']==cancer].loc[ix, 'PFS'], \n",
    "                           var1 = state_scores[cancer][state], var1_name = state)\n",
    "        \n",
    "        for end in cph_dict.keys(): \n",
    "            \n",
    "            cph_concat[end] = pd.concat([cph_concat[end], cph_dict[end]['cph']])\n",
    "            \n",
    "    \n",
    "    for end in cph_concat.keys():\n",
    "        print(end)\n",
    "        \n",
    "        \n",
    "        plot_forest_cph(cph_concat[end], title = cancer + ' ' + end, xlabel = 'log10(HR) (95% CI) (n=' + str(cph_dict[end]['n']) + ')', sort_on_pvalues = True, nan_str = False,   \n",
    "                        fig_adjust = (5, 1), save = './outputs/survival/' + cancer + '/cph/' + end +'_chemokine_receptors.pdf')\n",
    "    \n",
    "        print(cph_dict[end]['na'])\n",
    "    print('\\n')\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All cancers multivariate CPH strafified by age and covariate = sex "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CD8B', 'CXCR6', 'CXCR3', 'CXCR2', 'CCR5', 'CCR10', 'IL12B', 'CD19', 'NK cells', 'CXCL16']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD8B\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.579\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.555\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.543\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 384 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CD8B</th>\n",
       "      <th>km</th>\n",
       "      <td>0.51</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>6.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>1.13</td>\n",
       "      <td>0.29</td>\n",
       "      <td>1.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CD8B & km &        0.506385 &  0.476708 &  1.068823 \\\\\n",
       "    & rank &        6.035947 &  0.014017 &  6.156635 \\\\\n",
       "Sex & km &        0.001369 &  0.970485 &  0.043222 \\\\\n",
       "    & rank &        1.130427 &  0.287684 &  1.797442 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 384 total observations, 237 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "           test_statistic    p  -log2(p)\n",
       "CD8B km              0.51 0.48      1.07\n",
       "     rank            6.04 0.01      6.16\n",
       "Sex  km              0.00 0.97      0.04\n",
       "     rank            1.13 0.29      1.80"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CD8B' failed the non-proportional test: p-value is 0.0140.\n",
      "\n",
      "Concordance index: 0.537\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.524\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.523\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.561\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.563\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.628\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CD8B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CXCR6\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.572\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.566\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.574\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.573\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.512\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.499\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.513\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.574\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.565\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.628\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CXCR6': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CXCR3\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.566\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.584\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.5\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.531\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.511\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.538\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.569\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.624\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.568\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CXCR3': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CXCR2\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.546\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.517\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.521\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.547\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.541\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.522\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CXCR2': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CCR5\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.57\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.57\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.572\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.546\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.519\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.506\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.506\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.554\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.564\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.627\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CCR5': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CCR10\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.577\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.533\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.581\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.574\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.505\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.559\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 386 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR10</th>\n",
       "      <th>km</th>\n",
       "      <td>1.09</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>7.72</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>2.66</td>\n",
       "      <td>0.10</td>\n",
       "      <td>3.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR10 & km &        1.088798 &  0.296738 &  1.752740 \\\\\n",
       "    & rank &        7.716973 &  0.005470 &  7.514137 \\\\\n",
       "Sex & km &        0.001722 &  0.966902 &  0.048558 \\\\\n",
       "    & rank &        2.664735 &  0.102595 &  3.284969 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 386 total observations, 202 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic    p  -log2(p)\n",
       "CCR10 km              1.09 0.30      1.75\n",
       "      rank            7.72 0.01      7.51\n",
       "Sex   km              0.00 0.97      0.05\n",
       "      rank            2.66 0.10      3.28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR10' failed the non-proportional test: p-value is 0.0055.\n",
      "\n",
      "Concordance index: 0.58\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 405 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR10</th>\n",
       "      <th>km</th>\n",
       "      <td>3.67</td>\n",
       "      <td>0.06</td>\n",
       "      <td>4.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>9.56</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>0.05</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>0.56</td>\n",
       "      <td>0.46</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR10 & km &        3.673041 &  0.055299 &  4.176597 \\\\\n",
       "    & rank &        9.555014 &  0.001994 &  8.970090 \\\\\n",
       "Sex & km &        0.049003 &  0.824807 &  0.277872 \\\\\n",
       "    & rank &        0.555453 &  0.456098 &  1.132583 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 405 total observations, 228 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic      p  -log2(p)\n",
       "CCR10 km              3.67   0.06      4.18\n",
       "      rank            9.56 <0.005      8.97\n",
       "Sex   km              0.05   0.82      0.28\n",
       "      rank            0.56   0.46      1.13"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR10' failed the non-proportional test: p-value is 0.0020.\n",
      "\n",
      "Concordance index: 0.541\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 406 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CCR10</th>\n",
       "      <th>km</th>\n",
       "      <td>1.59</td>\n",
       "      <td>0.21</td>\n",
       "      <td>2.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>13.58</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>12.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>0.49</td>\n",
       "      <td>0.48</td>\n",
       "      <td>1.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>2.77</td>\n",
       "      <td>0.10</td>\n",
       "      <td>3.38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &   -log2(p) \\\\\n",
       "\\midrule\n",
       "CCR10 & km &        1.587966 &  0.207617 &   2.268003 \\\\\n",
       "    & rank &       13.583650 &  0.000228 &  12.097640 \\\\\n",
       "Sex & km &        0.492012 &  0.483031 &   1.049812 \\\\\n",
       "    & rank &        2.774143 &  0.095798 &   3.383862 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 406 total observations, 178 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "            test_statistic      p  -log2(p)\n",
       "CCR10 km              1.59   0.21      2.27\n",
       "      rank           13.58 <0.005     12.10\n",
       "Sex   km              0.49   0.48      1.05\n",
       "      rank            2.77   0.10      3.38"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CCR10' failed the non-proportional test: p-value is 0.0002.\n",
      "\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CCR10': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "IL12B\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.587\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.601\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.556\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.575\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.55\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.512\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.503\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.534\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.56\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.553\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.537\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.591\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.539\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'IL12B': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CD19\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.575\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.581\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.622\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.613\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.558\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.538\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.51\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.567\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.532\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.525\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.529\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.586\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.549\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CD19': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "NK cells\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.57\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.552\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.566\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.551\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.484\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.503\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.523\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.503\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.522\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.535\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.545\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.533\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.571\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.578\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.587\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.542\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'NK cells': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "CXCL16\n",
      "BRCA\n",
      "Fitting: OS ...\n",
      "Number of events: 140.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 191.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 1070 total...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCL16</th>\n",
       "      <th>km</th>\n",
       "      <td>7.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>0.12</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>0.38</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>1.59</td>\n",
       "      <td>0.21</td>\n",
       "      <td>2.27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCL16 & km &        7.029253 &  0.008019 &  6.962384 \\\\\n",
       "    & rank &        0.121586 &  0.727321 &  0.459337 \\\\\n",
       "Sex & km &        0.382552 &  0.536241 &  0.899047 \\\\\n",
       "    & rank &        1.594997 &  0.206614 &  2.274992 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 1070 total observations, 879 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "             test_statistic    p  -log2(p)\n",
       "CXCL16 km              7.03 0.01      6.96\n",
       "       rank            0.12 0.73      0.46\n",
       "Sex    km              0.38 0.54      0.90\n",
       "       rank            1.59 0.21      2.27"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCL16' failed the non-proportional test: p-value is 0.0080.\n",
      "\n",
      "Concordance index: 0.518\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "1070\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "1070\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "HNSC\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.526\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 234.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "444\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "444\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 75, 'time': 1, 'event': 0}\n",
      "LUAD\n",
      "Fitting: OS ...\n",
      "Number of events: 181.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.562\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 256.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 498 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CXCL16</th>\n",
       "      <th>km</th>\n",
       "      <td>6.73</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>3.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>3.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Sex</th>\n",
       "      <th>km</th>\n",
       "      <td>3.01</td>\n",
       "      <td>0.08</td>\n",
       "      <td>3.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>0.08</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "    &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "CXCL16 & km &        6.725345 &  0.009505 &  6.717062 \\\\\n",
       "    & rank &        3.070863 &  0.079707 &  3.649147 \\\\\n",
       "Sex & km &        3.012072 &  0.082647 &  3.596901 \\\\\n",
       "    & rank &        0.079777 &  0.777600 &  0.362901 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 498 total observations, 242 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "             test_statistic    p  -log2(p)\n",
       "CXCL16 km              6.73 0.01      6.72\n",
       "       rank            3.07 0.08      3.65\n",
       "Sex    km              3.01 0.08      3.60\n",
       "       rank            0.08 0.78      0.36"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'CXCL16' failed the non-proportional test: p-value is 0.0095.\n",
      "\n",
      "Concordance index: 0.559\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "498\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "PFS\n",
      "498\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 8, 'time': 9, 'event': 0}\n",
      "LUSC\n",
      "Fitting: OS ...\n",
      "Number of events: 210.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.548\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 250.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.527\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "491\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "492\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 4, 'time': 5, 'event': 0}\n",
      "STAD\n",
      "Fitting: OS ...\n",
      "Number of events: 147.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.514\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 184.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.536\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "384\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 6, 'event': 0}\n",
      "PFS\n",
      "386\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 25, 'time': 4, 'event': 0}\n",
      "KIRC\n",
      "Fitting: OS ...\n",
      "Number of events: 174.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.508\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 221.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "530\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 0, 'event': 0}\n",
      "PFS\n",
      "528\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 3, 'time': 2, 'event': 0}\n",
      "BLCA\n",
      "Fitting: OS ...\n",
      "Number of events: 177.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.53\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 228.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.528\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "405\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "406\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 2, 'time': 0, 'event': 0}\n",
      "LIHC\n",
      "Fitting: OS ...\n",
      "Number of events: 116.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.52\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 208.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.516\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "346\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "PFS\n",
      "346\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 24, 'time': 1, 'event': 0}\n",
      "SKCM\n",
      "Fitting: OS ...\n",
      "Number of events: 189.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.573\n",
      "Finished fitting: OS!\n",
      "\n",
      "Fitting: PFS ...\n",
      "Number of events: 287.0\n",
      "Proportional hazard assumption looks okay.\n",
      "Concordance index: 0.54\n",
      "Finished fitting: PFS!\n",
      "\n",
      "OS\n",
      "408\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 9, 'event': 7}\n",
      "PFS\n",
      "409\n",
      "{'CXCL16': 0, 'Sex': 0, 'Stage (versus 0 & I)': 52, 'time': 8, 'event': 7}\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.mkdir('./outputs/survival/multivariate_all_cancers/')\n",
    "for state in states:\n",
    "    print(state)\n",
    "    \n",
    "    cph_concat = {}\n",
    "    cph_concat['OS'] = pd.DataFrame()\n",
    "    cph_concat['PFS'] = pd.DataFrame()\n",
    "    \n",
    "    for cancer in cancers:\n",
    "        print(cancer)\n",
    "        ix = state_scores[cancer][state].index.values\n",
    "        \n",
    "        cph_dict = fit_cph(end1_name = 'OS', end1_time = TCGA_sample_info.loc[ix,'OS.time'], end1_event = TCGA_sample_info.loc[ix, 'OS'], \n",
    "                           end2_name = 'PFS', end2_time = TCGA_survival.loc[ix, 'PFS.time'], end2_event = TCGA_survival.loc[ix, 'PFS'], \n",
    "                           var1 = state_scores[cancer][state], var1_name = state, \n",
    "                           var2 = TCGA_sample_info.loc[ix, 'sex'], var2_name = 'Sex', \n",
    "                           #var3 =  TCGA_sample_info.loc[ix, 'age_at_initial_pathologic_diagnosis'], var3_name = 'Age', \n",
    "                           dummy_var1 = TCGA_sample_info.loc[ix,'tumor_stage'], dummy_var1_name = ('Stage (versus 0 & I)', 'Stage 0 & I'), \n",
    "                           advice = False, strata = ['Stage (versus 0 & I)'], drop = ['Sex'])\n",
    "        \n",
    "        for end in cph_dict.keys():\n",
    "            print(end)\n",
    "            print(cph_dict[end]['n'])\n",
    "            print(cph_dict[end]['na'])\n",
    "            \n",
    "            cph_dict[end]['cph'].index = cancer + ' ' + cph_dict[end]['cph'].index \n",
    "            cph_concat[end] = pd.concat([cph_concat[end], cph_dict[end]['cph']])\n",
    "        \n",
    "    for end in cph_concat.keys():\n",
    "\n",
    "        plot_forest_cph(cph_concat[end], title = end, xlabel = 'log10(HR) (95% CI)', sort_on_pvalues = True, nan_str = False, \n",
    "                        fig_adjust = (5, 1.25), save = './outputs/survival/multivariate_all_cancers/' + end + '_' + state +'.pdf')\n",
    "    \n",
    "    print('\\n')\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Kaplan Meiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ['CD8B', 'CXCR6', 'CXCR3', 'CXCR2', 'CCR5', 'CCR10', 'IL12B', 'CXCL16', 'NK cells', 'CD19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA\n",
      "Number of patients in 0_33 group CD8B: 365\n",
      "-3.36906952828317 -0.3797748698479904 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 365\n",
      "0.46397804076015414 2.925674533481078 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 365\n",
      "-3.315627990370191 -0.4318423731465461 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 365\n",
      "0.4596873084543077 2.7074374084166895 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 365\n",
      "-3.5641891989573553 -0.38450408689199633 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 365\n",
      "0.4068573531684912 2.510956860210585 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 365\n",
      "-3.1549186452901568 -0.3048504812409999 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 365\n",
      "0.42487742543984863 5.069117017843321 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 365\n",
      "-4.024744375427019 -0.4232818580538839 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 365\n",
      "0.46663258730890633 2.8339548726892962 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 365\n",
      "-3.235943817821323 -0.4568297614025395 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 365\n",
      "0.39816132184032244 3.835429571433662 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 365\n",
      "-1.6520812867299093 -0.5097525289525385 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 365\n",
      "0.49030754924879655 2.6753645338757712 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 365\n",
      "-3.7626070419920037 -0.39305688762880464 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 365\n",
      "0.3663419642323125 3.6772958962333044 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 365\n",
      "-1.8017219674591667 -0.48697892554052935 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 365\n",
      "0.22621096278501895 3.816747305816859 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 365\n",
      "-1.456117833580244 -0.600547950455865 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 365\n",
      "0.39260293679607083 3.2396608959394246 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 76.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 62.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 83.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 84.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 83.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 61.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 89.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 61.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 86.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 50.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 58.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 82.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 50.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 90.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 54.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 56.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 45.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 61.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 39.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 61.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 40.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 61.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 46.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 64.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 39.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 42.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 64.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 42.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 62.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 43.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 65.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 37.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 67.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 42.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "HNSC\n",
      "Number of patients in 0_33 group CD8B: 173\n",
      "-2.97037139116653 -0.38265455530857445 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 174\n",
      "0.46140930059853563 2.2542303236346446 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 173\n",
      "-3.024227093232007 -0.3746583033923622 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 174\n",
      "0.46280862482798923 2.2616497022107174 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 173\n",
      "-3.3440505729129173 -0.4056215473672137 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 174\n",
      "0.4387877366814668 2.450275306729162 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 174\n",
      "-3.2793275181086043 -0.4564430526835954 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 174\n",
      "0.46308627143145115 2.869290973064007 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 173\n",
      "-3.0010101986355964 -0.4189344835838807 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 174\n",
      "0.5034338532093859 2.2445278750539877 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 173\n",
      "-2.651476137580114 -0.4477864155055813 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 174\n",
      "0.2659839843411157 3.3332363836024967 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 174\n",
      "-1.764885922843542 -0.5118094307381548 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 174\n",
      "0.39319361170304507 2.77687413776051 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 173\n",
      "-3.3576478394249474 -0.3781102828539479 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 174\n",
      "0.41311324651789244 2.614518097965012 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 173\n",
      "-1.7922846548772293 -0.49565008116537973 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 174\n",
      "0.32227239506347605 3.378376620337728 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 174\n",
      "-1.8527027843067194 -0.5044304111940099 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 174\n",
      "0.32601879724151794 3.594959470862091 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 102.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 104.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 105.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 80.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 94.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 94.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 108.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 87.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 102.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 95.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 81.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 89.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 92.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 99.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 91.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 114.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 81.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 65.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 85.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 61.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 84.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 62.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 76.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 87.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 81.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 57.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 81.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 63.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 71.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 81.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 73.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 97.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 347 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>4.75</td>\n",
       "      <td>0.03</td>\n",
       "      <td>5.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>6.69</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        4.752712 &  0.029252 &  5.095313 \\\\\n",
       "   & rank &        6.685409 &  0.009721 &  6.684750 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 347 total observations, 195 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic    p  -log2(p)\n",
       "ix km              4.75 0.03      5.10\n",
       "   rank            6.69 0.01      6.68"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0097.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LUAD\n",
      "Number of patients in 0_33 group CD8B: 172\n",
      "-3.2248854660298383 -0.4277101905591369 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 172\n",
      "0.4429544259680269 2.6578710336133984 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 172\n",
      "-4.425468462500174 -0.4722008367657788 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 172\n",
      "0.4310071209453733 2.8131519783741443 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 172\n",
      "-3.6289326876537413 -0.4443338040048026 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 172\n",
      "0.45593869946278964 3.2746359598045602 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 172\n",
      "-3.675206400831169 -0.456667151952768 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 172\n",
      "0.4555506391631498 2.795635359204796 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 172\n",
      "-3.128991669965957 -0.4260676442518078 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 172\n",
      "0.5140528168589494 2.3298730094224602 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 172\n",
      "-3.282857894966253 -0.4755681221196202 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 172\n",
      "0.25250863012341335 3.5000735272064047 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 172\n",
      "-2.6385309906221255 -0.37664378048143676 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 172\n",
      "0.48664889687703466 4.171498298955648 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 172\n",
      "-4.493273536016876 -0.24923206650004756 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 172\n",
      "0.47634508992495234 2.7530969310907345 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 172\n",
      "-2.1464129840609254 -0.46515311989092956 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 172\n",
      "0.3516540702012383 3.3972606626558766 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 172\n",
      "-2.8541924807368213 -0.3954694442225791 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 172\n",
      "0.5540466759666486 2.243825968427779 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 95.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 87.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 100.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 96.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 92.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 77.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 98.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 78.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 338 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>9.86</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>9.26</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        9.863347 &  0.001686 &  9.212145 \\\\\n",
       "   & rank &        9.262726 &  0.002339 &  8.740111 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 338 total observations, 162 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic      p  -log2(p)\n",
       "ix km              9.86 <0.005      9.21\n",
       "   rank            9.26 <0.005      8.74"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0017.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 85.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 83.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 92.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 78.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 337 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>10.23</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>8.86</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        10.22607 &  0.001385 &  9.496217 \\\\\n",
       "   & rank &         8.85682 &  0.002920 &  8.419836 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 337 total observations, 167 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic      p  -log2(p)\n",
       "ix km             10.23 <0.005      9.50\n",
       "   rank            8.86 <0.005      8.42"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0014.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 94.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 98.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 80.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 99.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 78.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 338 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>8.91</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>8.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>9.74</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>9.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        8.912849 &  0.002832 &  8.464110 \\\\\n",
       "   & rank &        9.744005 &  0.001799 &  9.118521 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 338 total observations, 161 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic      p  -log2(p)\n",
       "ix km              8.91 <0.005      8.46\n",
       "   rank            9.74 <0.005      9.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0018.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "OS:  CD8B\n",
      "Number of events ix0: 68.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 59.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 75.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 48.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 72.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 66.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 46.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 77.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 48.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 59.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 59.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 76.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 54.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 337 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>7.92</td>\n",
       "      <td>&lt;0.005</td>\n",
       "      <td>7.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>5.44</td>\n",
       "      <td>0.02</td>\n",
       "      <td>5.67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        7.919254 &  0.004891 &  7.675612 \\\\\n",
       "   & rank &        5.444397 &  0.019631 &  5.670709 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 337 total observations, 207 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic      p  -log2(p)\n",
       "ix km              7.92 <0.005      7.68\n",
       "   rank            5.44   0.02      5.67"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0049.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 69.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 54.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 67.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 57.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 48.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LUSC\n",
      "Number of patients in 0_33 group CD8B: 167\n",
      "-3.0781257009309337 -0.3944814966347303 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 167\n",
      "0.40602096508853586 4.0434061256400335 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 167\n",
      "-3.861502447603206 -0.40194554578002517 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 167\n",
      "0.4955615195910406 2.8807531803833544 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 167\n",
      "-3.3073866136032524 -0.41410168172087525 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 167\n",
      "0.4891271649970565 2.223674994135732 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 167\n",
      "-2.7818337505987896 -0.49763006418979455 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 167\n",
      "0.4109608610977626 3.088772107527189 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 167\n",
      "-3.3426888703080277 -0.39156744599301313 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 167\n",
      "0.4536803971388816 1.9860520512225226 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 167\n",
      "-3.430342638108323 -0.47398527642252763 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 167\n",
      "0.36366814979324874 3.893769125719743 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 167\n",
      "-1.9784188012627983 -0.4612512576566223 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 167\n",
      "0.4721851785461162 3.5164804912958556 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 167\n",
      "-3.4073364354822955 -0.38241412437259426 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 167\n",
      "0.506531467399399 3.423719046149646 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 167\n",
      "-1.9635466032939028 -0.41113243045997766 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 167\n",
      "0.4025229600324127 3.0639998507324524 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 167\n",
      "-2.6309967043203213 -0.40018489222075293 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 167\n",
      "0.5513401918093803 3.412756147083036 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 88.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 81.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 88.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 82.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 88.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 84.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 90.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 82.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 85.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 72.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 83.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 88.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 71.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 65.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 70.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 70.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 75.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 71.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 67.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 75.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 70.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 73.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 62.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 64.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 75.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 70.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "STAD\n",
      "Number of patients in 0_33 group CD8B: 138\n",
      "-2.7625726423749075 -0.3522977782167662 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 139\n",
      "0.4734044230673339 3.2096084348929677 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 138\n",
      "-3.1737549532645764 -0.404103834573455 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 139\n",
      "0.545636235484665 2.4162776792004177 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 138\n",
      "-3.111636049630288 -0.4164287045298943 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 139\n",
      "0.4005727648798718 2.6576519826457035 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 138\n",
      "-2.6605001627511915 -0.5500464084967381 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 139\n",
      "0.3514641379483848 3.0520093684388145 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 138\n",
      "-3.558896629093519 -0.3937853171192726 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 139\n",
      "0.5053232829253496 2.3249045606727514 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 139\n",
      "-4.369793528297351 -0.46512358699338946 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 139\n",
      "0.37532939267837917 3.193618245775288 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 138\n",
      "-1.8053022194783424 -0.4576973318691126 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 139\n",
      "0.41683516141385063 3.4582044919028365 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 138\n",
      "-3.6238811242424944 -0.27704862836002236 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 139\n",
      "0.5067811419485425 3.261485267085993 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 138\n",
      "-2.3646012250972954 -0.34059509112909436 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 139\n",
      "0.38967312621981276 2.668141924276439 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 138\n",
      "-2.4719194852541957 -0.43175231887050786 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 139\n",
      "0.4545352965888912 3.0002362338983795 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 58.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 70.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 60.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 65.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 65.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 67.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 61.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 66.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 52.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 57.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 66.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 66.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 65.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 56.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 70.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 44.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 55.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 58.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 49.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 53.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 49.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 55.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 50.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 52.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 39.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 49.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 53.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 53.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 54.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 51.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 42.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 57.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "KIRC\n",
      "Number of patients in 0_33 group CD8B: 178\n",
      "-3.1230594589096885 -0.3829285633382899 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 178\n",
      "0.45187592303000856 2.1744476557687737 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 178\n",
      "-4.384645716382011 -0.3354296902706975 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 178\n",
      "0.47996960859937565 2.4158867788482423 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 178\n",
      "-4.410588620391014 -0.3322226005208838 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 178\n",
      "0.4015404534548301 2.329655297883329 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 178\n",
      "-3.0235936239315087 -0.378269908087828 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 178\n",
      "0.402116910717589 4.688037612701721 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 178\n",
      "-4.168581701745398 -0.3256292806306565 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 178\n",
      "0.4434555402482463 2.2849591520821972 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 178\n",
      "-4.380003615842839 -0.36348663854990315 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 178\n",
      "0.44342076242401285 4.618570230367682 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 178\n",
      "-1.8020665905924853 -0.45560334117035783 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 178\n",
      "0.41253642048133116 2.9980910648604255 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 178\n",
      "-3.2183701076117908 -0.3999404391433873 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 178\n",
      "0.45400369768164367 2.2705622481197367 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 178\n",
      "-3.256918076599333 -0.3055767757162421 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 178\n",
      "0.40518314357222573 2.368206281052977 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 178\n",
      "-1.4809529032276343 -0.5105271182772618 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 178\n",
      "0.23644528547703772 4.568953322300179 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 70.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 80.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 87.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 83.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 68.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 70.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 65.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 90.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 76.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 71.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 72.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 82.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 89.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 67.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 67.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 83.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 58.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 68.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 48.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 68.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 48.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 76.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 66.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 53.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 56.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 66.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 44.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 60.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 60.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 55.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 67.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 71.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 46.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "BLCA\n",
      "Number of patients in 0_33 group CD8B: 136\n",
      "-2.6703110364636395 -0.40449224286772073 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 136\n",
      "0.39603818335114804 3.361467507383329 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 136\n",
      "-2.833671385682469 -0.4215719772036859 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 136\n",
      "0.4393245982408574 2.6556128322490875 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 136\n",
      "-2.751364669017636 -0.4588301310403743 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 136\n",
      "0.42809531153832664 2.4550393781437205 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 136\n",
      "-2.7137890209667983 -0.4944867224424607 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 136\n",
      "0.4637327476644147 2.7786469656092363 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 136\n",
      "-3.6289808367514267 -0.4585115211661863 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 136\n",
      "0.5226663998202964 2.4416541086960724 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 136\n",
      "-2.2979156851626397 -0.4724966062505239 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 136\n",
      "0.3374732134489147 3.458131004170665 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 136\n",
      "-1.236328564930262 -0.6137642832645307 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 136\n",
      "0.3714385982209177 3.370450599666582 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 136\n",
      "-4.099750802536314 -0.3754153766279661 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 136\n",
      "0.4376552185763539 2.8682833152732194 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 136\n",
      "-1.6863645622660006 -0.5022933489010677 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 136\n",
      "0.3358431486345217 2.6985201460728474 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 136\n",
      "-1.388626011752736 -0.6224550083492079 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 136\n",
      "0.37048934682181406 3.387338266777727 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 77.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 76.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "The ``p_value_threshold`` is set at 0.01. Even under the null hypothesis of no violations, some\n",
      "covariates will be below the threshold by chance. This is compounded when there are many covariates.\n",
      "Similarly, when there are lots of observations, even minor deviances from the proportional hazard\n",
      "assumption will be flagged.\n",
      "\n",
      "With that in mind, it's best to use a combination of statistical tests and visual tests to determine\n",
      "the most serious violations. Produce visual plots using ``check_assumptions(..., show_plots=True)``\n",
      "and looking for non-constant lines. See link [A] below for a full example.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>null_distribution</th>\n",
       "      <td>chi squared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>degrees_of_freedom</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>&lt;lifelines.CoxPHFitter: fitted with 272 total ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>test_name</th>\n",
       "      <td>proportional_hazard_test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>test_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>-log2(p)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ix</th>\n",
       "      <th>km</th>\n",
       "      <td>5.26</td>\n",
       "      <td>0.02</td>\n",
       "      <td>5.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td>5.92</td>\n",
       "      <td>0.01</td>\n",
       "      <td>6.06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/latex": [
       "\\begin{tabular}{llrrr}\n",
       "\\toprule\n",
       "   &      &  test\\_statistic &         p &  -log2(p) \\\\\n",
       "\\midrule\n",
       "ix & km &        5.264182 &  0.021769 &  5.521604 \\\\\n",
       "   & rank &        5.922443 &  0.014949 &  6.063786 \\\\\n",
       "\\bottomrule\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "<lifelines.StatisticalResult: proportional_hazard_test>\n",
       " null_distribution = chi squared\n",
       "degrees_of_freedom = 1\n",
       "             model = <lifelines.CoxPHFitter: fitted with 272 total observations, 124 right-censored observations>\n",
       "         test_name = proportional_hazard_test\n",
       "\n",
       "---\n",
       "         test_statistic    p  -log2(p)\n",
       "ix km              5.26 0.02      5.52\n",
       "   rank            5.92 0.01      6.06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "1. Variable 'ix' failed the non-proportional test: p-value is 0.0149.\n",
      "\n",
      "   Advice: with so few unique values (only 2), you can include `strata=['ix', ...]` in the call in\n",
      "`.fit`. See documentation in link [E] below.\n",
      "\n",
      "---\n",
      "[A]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html\n",
      "[B]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Bin-variable-and-stratify-on-it\n",
      "[C]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Introduce-time-varying-covariates\n",
      "[D]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Modify-the-functional-form\n",
      "[E]  https://lifelines.readthedocs.io/en/latest/jupyter_notebooks/Proportional%20hazard%20assumption.html#Stratification\n",
      "\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 77.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 89.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 76.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 71.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 84.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 85.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 74.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 74.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 79.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 60.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 63.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 60.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 58.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 65.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 46.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 58.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 61.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 71.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 67.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 57.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 58.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 60.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 65.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 56.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 56.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 59.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "LIHC\n",
      "Number of patients in 0_33 group CD8B: 124\n",
      "-2.562240745172741 -0.4454036229687092 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 124\n",
      "0.28891938118310373 2.7425617747217323 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 124\n",
      "-3.625979736036324 -0.377589338672637 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 124\n",
      "0.44210078469100517 2.6729020450644234 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 124\n",
      "-2.735993456602109 -0.43625119467811224 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 124\n",
      "0.3968825765180518 2.7730913515607654 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 124\n",
      "-2.208629598140941 -0.4274737537906998 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 124\n",
      "0.36896992876788304 3.658004443102418 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 124\n",
      "-2.9979151101248456 -0.4332215102522103 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 124\n",
      "0.45181320512132067 2.7211139460612013 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 124\n",
      "-2.9064382643793265 -0.3426228477658914 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 124\n",
      "0.24211664465007715 4.379764117219231 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 124\n",
      "-1.3924803921077977 -0.7511291830651168 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 124\n",
      "0.4953981825035131 3.3770597930773496 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 124\n",
      "-2.6623628502162657 -0.4779689706864675 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 124\n",
      "0.39765028842841876 2.8901049436299235 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 124\n",
      "-1.9980928352011524 -0.43925211740244646 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 124\n",
      "0.3187947616160506 3.3752055994007515 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 124\n",
      "-1.3755790714137333 -0.6081346817298676 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 124\n",
      "0.2999538166575207 3.645631113287685 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 78.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 66.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 84.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 66.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 83.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 68.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 73.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 78.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 63.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 70.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 80.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 75.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 69.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 86.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 63.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 72.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 45.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 38.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 51.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 33.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 50.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 39.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 40.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 43.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 46.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 36.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 36.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 52.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 46.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 43.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 42.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 41.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 47.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 37.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 47.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 41.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SKCM\n",
      "Number of patients in 0_33 group CD8B: 157\n",
      "-2.453374779921947 -0.4607645500019852 \n",
      "\n",
      "Number of patients in 67-100 group CD8B: 157\n",
      "0.5677229809390804 2.143254260416845 \n",
      "\n",
      "Number of patients in 0_33 group CXCR6: 156\n",
      "-2.7124408245054377 -0.46599583814620577 \n",
      "\n",
      "Number of patients in 67-100 group CXCR6: 157\n",
      "0.4989105205610839 2.343921557909661 \n",
      "\n",
      "Number of patients in 0_33 group CXCR3: 156\n",
      "-2.6965762807828453 -0.4482878365516078 \n",
      "\n",
      "Number of patients in 67-100 group CXCR3: 157\n",
      "0.5530201407299012 2.6260083418574407 \n",
      "\n",
      "Number of patients in 0_33 group CXCR2: 156\n",
      "-1.8824648356842857 -0.48772638753170977 \n",
      "\n",
      "Number of patients in 67-100 group CXCR2: 157\n",
      "0.2757125325614847 3.2235985768883384 \n",
      "\n",
      "Number of patients in 0_33 group CCR5: 156\n",
      "-3.07043344578503 -0.4634333001316263 \n",
      "\n",
      "Number of patients in 67-100 group CCR5: 157\n",
      "0.5382983364826226 2.501869422564624 \n",
      "\n",
      "Number of patients in 0_33 group CCR10: 157\n",
      "-2.925670632901845 -0.47601038310044114 \n",
      "\n",
      "Number of patients in 67-100 group CCR10: 157\n",
      "0.37224977544551835 3.405480251936486 \n",
      "\n",
      "Number of patients in 0_33 group IL12B: 156\n",
      "-1.3761014093578903 -0.6109816769305279 \n",
      "\n",
      "Number of patients in 67-100 group IL12B: 157\n",
      "0.41114519306894415 3.630767785033312 \n",
      "\n",
      "Number of patients in 0_33 group CXCL16: 157\n",
      "-3.2083049839750126 -0.3585502344280273 \n",
      "\n",
      "Number of patients in 67-100 group CXCL16: 157\n",
      "0.48841599632772387 3.3601987882449484 \n",
      "\n",
      "Number of patients in 0_33 group NK cells: 156\n",
      "-1.5977655019125967 -0.47479966309312704 \n",
      "\n",
      "Number of patients in 67-100 group NK cells: 157\n",
      "0.3562972722991786 2.479718688926681 \n",
      "\n",
      "Number of patients in 0_33 group CD19: 157\n",
      "-1.3684170286193669 -0.6783899013169968 \n",
      "\n",
      "Number of patients in 67-100 group CD19: 157\n",
      "0.4988878205021246 2.6564689056904935 \n",
      "\n",
      "PFS:  CD8B\n",
      "Number of events ix0: 118.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 101.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR6\n",
      "Number of events ix0: 118.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 103.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR3\n",
      "Number of events ix0: 119.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 101.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCR2\n",
      "Number of events ix0: 119.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 107.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR5\n",
      "Number of events ix0: 112.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 108.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CCR10\n",
      "Number of events ix0: 111.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 108.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  IL12B\n",
      "Number of events ix0: 113.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 103.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CXCL16\n",
      "Number of events ix0: 110.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 102.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  NK cells\n",
      "Number of events ix0: 110.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 102.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "PFS:  CD19\n",
      "Number of events ix0: 113.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 100.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD8B\n",
      "Number of events ix0: 92.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 52.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR6\n",
      "Number of events ix0: 91.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 50.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR3\n",
      "Number of events ix0: 88.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 54.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCR2\n",
      "Number of events ix0: 79.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 75.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR5\n",
      "Number of events ix0: 86.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 54.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CCR10\n",
      "Number of events ix0: 77.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 64.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  IL12B\n",
      "Number of events ix0: 83.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 60.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CXCL16\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 61.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  NK cells\n",
      "Number of events ix0: 80.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 58.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "OS:  CD19\n",
      "Number of events ix0: 81.0 Number of missing values ix0: 0\n",
      "Number of events ix67: 52.0 Number of missing values ix67: 0\n",
      "Proportional hazard assumption looks okay.\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "quantile = 1/3\n",
    "for cancer in cancers:\n",
    "    print(cancer)\n",
    "    os.mkdir('./outputs/survival/' + cancer + '/kaplan_meier/')\n",
    "    # dichotomize patients into botom 1/3 and top 1/3 groups using the state signature scores: \n",
    "    state_scores_0 = {}\n",
    "    state_scores_66 = {}\n",
    "\n",
    "    for state in states: \n",
    "        cmask = (state_scores[cancer][state] <= state_scores[cancer][state].quantile(q = quantile))\n",
    "        state_scores_0[state] = state_scores[cancer][state][cmask]\n",
    "        print('Number of patients in 0_33 group '+ state + ':', len(state_scores_0[state]))\n",
    "        print(state_scores_0[state].min(), state_scores_0[state].max(), '\\n')\n",
    "    \n",
    "        cmask = (state_scores[cancer][state] >= state_scores[cancer][state].quantile(q = 2*quantile))\n",
    "        state_scores_66[state] = state_scores[cancer][state][cmask]\n",
    "        print('Number of patients in 67-100 group '+ state + ':', len(state_scores_66[state]))\n",
    "        print(state_scores_66[state].min(), state_scores_66[state].max(), '\\n')\n",
    "    \n",
    "    # Extract OS_time/event and PFS_time/event: \n",
    "\n",
    "    OS_time = TCGA_sample_info[TCGA_sample_info['type']==cancer]['OS.time']\n",
    "    OS_status = TCGA_sample_info[TCGA_sample_info['type']==cancer]['OS']\n",
    "    PFS_time =TCGA_survival[TCGA_survival['type']==cancer]['PFS.time']\n",
    "    PFS_status = TCGA_survival[TCGA_survival['type']==cancer]['PFS']\n",
    "\n",
    "    # Fit KaplanMeier PFS and plot:  \n",
    "    for state in states:\n",
    "        print('PFS: ', state)\n",
    "        ix0 = state_scores_0[state].index.values \n",
    "        ix67 = state_scores_66[state].index.values \n",
    "        \n",
    "        cph_test_0 = pd.DataFrame({'PFS_time': PFS_time[ix0].dropna()/365, \n",
    "                                   'ix': 0, \n",
    "                                   'PFS_event': PFS_status[PFS_time[ix0].dropna().index.values].dropna()}) # here you can do the last .dropna() because you fill ix with dummy 0 \n",
    "        \n",
    "        cph_test_67 = pd.DataFrame({'PFS_time': PFS_time[ix67].dropna()/365, \n",
    "                                    'ix': 1, \n",
    "                                    'PFS_event': PFS_status[PFS_time[ix67].dropna().index.values].dropna()}) # here you can do the last.dropna() because you fill ix with dummy 1\n",
    "        \n",
    "        print('Number of events ix0:', PFS_status[PFS_time[ix0].dropna().index.values].dropna().sum(), 'Number of missing values ix0:', PFS_status[PFS_time[ix0].dropna().index.values].isna().sum())\n",
    "        print('Number of events ix67:', PFS_status[PFS_time[ix67].dropna().index.values].dropna().sum(), 'Number of missing values ix67:', PFS_status[PFS_time[ix67].dropna().index.values].isna().sum())\n",
    "        \n",
    "        cph_test = pd.concat([cph_test_0, cph_test_67]) \n",
    "        \n",
    "        cph = CoxPHFitter()\n",
    "        cph.fit(cph_test, duration_col='PFS_time', event_col='PFS_event')\n",
    "        cph.check_assumptions(cph_test)\n",
    "        \n",
    "        kmf_0 = KaplanMeierFitter() \n",
    "        kmf_67 = KaplanMeierFitter()\n",
    "\n",
    "        kmf_0.fit(PFS_time[ix0].dropna()/365, \n",
    "                  PFS_status[PFS_time[ix0].dropna().index.values].dropna(), \n",
    "                  timeline = np.arange(0, 10, 0.005),\n",
    "                  label = 'Low (n=' + str(len(PFS_time[ix0].dropna())) + ')')\n",
    "        ax = kmf_0.plot(ci_show = False, color = 'black', show_censors = True, censor_styles={'marker': 2, 'ms':5, 'lw':0.5})\n",
    "\n",
    "        \n",
    "        kmf_67.fit(PFS_time[ix67].dropna()/365, \n",
    "                   PFS_status[PFS_time[ix67].dropna().index.values].dropna(),\n",
    "                   timeline = np.arange(0, 10, 0.005),\n",
    "                   label = 'High (n=' + str(len(PFS_time[ix67].dropna())) + ')')\n",
    "        ax = kmf_67.plot(ci_show = False, color = 'red', show_censors = True, censor_styles={'marker': 2, 'ms':5, 'lw':0.5})\n",
    "        \n",
    "        ax.text(0.01, 0.01, 'HR '+str(round(cph.summary.loc['ix', 'exp(coef)'], 2))+' ('+str(round(cph.summary.loc['ix', 'exp(coef) lower 95%'],2))+'-'+str(round(cph.summary.loc['ix', 'exp(coef) upper 95%'],2))+')'+'\\n'+'p='+str(round(cph.summary.loc['ix', 'p'], 10)),\n",
    "                ha='left', va='bottom', transform=ax.transAxes)\n",
    "            \n",
    "        #add_at_risk_counts(kmf_0, kmf_67, ax=ax)\n",
    "        plt.title(cancer + ' PFS ' + state)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/kaplan_meier/PFS_km_33_' + state + '.pdf', dpi = 400)\n",
    "        plt.close()\n",
    "    \n",
    "    # to make OS and PFS completely separate initialize a new for loop: \n",
    "    for state in states:\n",
    "        print('OS: ', state)\n",
    "        ix0 = state_scores_0[state].index.values \n",
    "        ix67 = state_scores_66[state].index.values \n",
    "        \n",
    "        cph_test_0 = pd.DataFrame({'OS_time': OS_time[ix0].dropna()/365, \n",
    "                                    'ix': 0, \n",
    "                                    'OS_event': OS_status[OS_time[ix0].dropna().index.values].dropna()}) # here you can do the last .dropna() because you fill ix with dummy 0 \n",
    "        \n",
    "        cph_test_67 = pd.DataFrame({'OS_time': OS_time[ix67].dropna()/365, \n",
    "                                    'ix': 1, \n",
    "                                    'OS_event': OS_status[OS_time[ix67].dropna().index.values].dropna()}) # here you can do the last .dropna() because you fill ix with dummy 1 \n",
    "        \n",
    "        print('Number of events ix0:', OS_status[OS_time[ix0].dropna().index.values].dropna().sum(), 'Number of missing values ix0:', OS_status[OS_time[ix0].dropna().index.values].isna().sum())\n",
    "        print('Number of events ix67:', OS_status[OS_time[ix67].dropna().index.values].dropna().sum(), 'Number of missing values ix67:', OS_status[OS_time[ix67].dropna().index.values].isna().sum())\n",
    "        \n",
    "        cph_test = pd.concat([cph_test_0, cph_test_67]) \n",
    "        \n",
    "        cph = CoxPHFitter()\n",
    "        cph.fit(cph_test, duration_col='OS_time', event_col='OS_event')\n",
    "        cph.check_assumptions(cph_test)\n",
    "        \n",
    "        kmf_0 = KaplanMeierFitter() \n",
    "        kmf_67 = KaplanMeierFitter()\n",
    "\n",
    "        kmf_0.fit(OS_time[ix0].dropna()/365, \n",
    "                  OS_status[OS_time[ix0].dropna().index.values].dropna(),\n",
    "                  timeline = np.arange(0, 10, 0.005),\n",
    "                  label = 'Low (n=' + str(len(OS_time[ix0].dropna())) + ')')\n",
    "        ax = kmf_0.plot(ci_show = False, color = 'black', show_censors = True, censor_styles={'marker': 2, 'ms':5, 'lw':0.5})\n",
    "        \n",
    "        kmf_67.fit(OS_time[ix67].dropna()/365, \n",
    "                   OS_status[OS_time[ix67].dropna().index.values].dropna(),\n",
    "                   timeline = np.arange(0, 10, 0.005),\n",
    "                   label = 'High (n=' + str(len(OS_time[ix67].dropna())) + ')')\n",
    "        ax = kmf_67.plot(ci_show = False, color = 'red', show_censors = True, censor_styles={'marker': 2, 'ms':5, 'lw':0.5})\n",
    "        \n",
    "        ax.text(0.01, 0.01, 'HR '+str(round(cph.summary.loc['ix', 'exp(coef)'], 2))+' ('+str(round(cph.summary.loc['ix', 'exp(coef) lower 95%'],2))+'-'+str(round(cph.summary.loc['ix', 'exp(coef) upper 95%'],2))+')'+'\\n'+'p='+str(round(cph.summary.loc['ix', 'p'], 10)),\n",
    "                ha='left', va='bottom', transform=ax.transAxes)\n",
    "        \n",
    "        #add_at_risk_counts(kmf_0, kmf_67, ax=ax)\n",
    "        plt.title(cancer + ' OS ' + state)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig('./outputs/survival/' + cancer + '/kaplan_meier/OS_km_33_' + state + '.pdf', dpi = 400)\n",
    "        plt.close()\n",
    "        \n",
    "    print('\\n')\n",
    "    print('\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
